diff --git a/runtime/arch/arm64/asm_support_arm64.S b/runtime/arch/arm64/asm_support_arm64.S
index 86c18d0..c264237 100644
--- a/runtime/arch/arm64/asm_support_arm64.S
+++ b/runtime/arch/arm64/asm_support_arm64.S
@@ -43,25 +43,49 @@
 #define xSUSPEND x21
 
 .macro LOAD_PC_REL_ADDRESS reg, symbol
+#if defined(__APPLE__)
+    adr \reg, _\symbol
+#else
     adr \reg, \symbol
+#endif
 .endm
 
 .macro CALL_SYMBOL symbol
+#if defined(__APPLE__)
+    bl _\symbol
+#else
     bl \symbol
+#endif
 .endm
 
 .macro BRANCH_SYMBOL symbol
+#if defined(__APPLE__)
+    b _\symbol
+#else
     b \symbol
+#endif
 .endm
 
 .macro BRANCH_SYMBOL_CBZ reg, symbol
+#if defined(__APPLE__)
+    cbz \reg, _\symbol
+#else
     cbz \reg, \symbol
+#endif
 .endm
 
 .macro BRANCH_SYMBOL_NE symbol
+#if defined(__APPLE__)
+    b.ne _\symbol
+#else
     b.ne \symbol
+#endif
 .endm
 
 .macro BRANCH_SYMBOL_EQ symbol
+#if defined(__APPLE__)
+    b.eq _\symbol
+#else
     b.eq \symbol
+#endif
 .endm
@@ -105,19 +129,28 @@
 .endm
 
 .macro ENTRY_ALIGNED name, alignment
+#if defined(__APPLE__)
+    .private_extern _\name
+    .global _\name
+    .balign \alignment
+_\name:
+#else
     .type \name, #function
     .hidden \name  // Hide this as a global symbol, so we do not incur plt calls.
     .global \name
     .balign \alignment
 \name:
+#endif
     .cfi_startproc
 .endm
 
 .macro ENTRY name
     ENTRY_ALIGNED \name, 16
 .endm
 
 .macro END name
     .cfi_endproc
+#if !defined(__APPLE__)
     .size \name, .-\name
+#endif
 .endm
@@ -230,8 +263,13 @@
 .macro LOAD_RUNTIME_INSTANCE reg
+#if defined(__APPLE__)
+    adrp \reg, __ZN3art7Runtime9instance_E@PAGE
+    ldr \reg, [\reg, __ZN3art7Runtime9instance_E@PAGEOFF]
+#else
 #if __has_feature(hwaddress_sanitizer)
     adrp \reg, :pg_hi21_nc:_ZN3art7Runtime9instance_E
 #else
     adrp \reg, _ZN3art7Runtime9instance_E
 #endif
     ldr \reg, [\reg, #:lo12:_ZN3art7Runtime9instance_E]
+#endif
 .endm
