diff --git a/runtime/arch/arm64/context_arm64.h b/runtime/arch/arm64/context_arm64.h
--- a/runtime/arch/arm64/context_arm64.h
+++ b/runtime/arch/arm64/context_arm64.h
@@ -76,7 +76,7 @@ class Arm64Context final : public Context {
   // Pointers to register locations, initialized to null or the specific registers below. We need
   // an additional one for the PC.
   uintptr_t* gprs_[kNumberOfXRegisters + 1];
-  uint64_t * fprs_[kNumberOfDRegisters];
+  uintptr_t* fprs_[kNumberOfDRegisters];
   // Hold values for sp, pc and arg0 if they are not located within a stack frame.
   uintptr_t sp_, pc_, arg0_;
 };
diff --git a/runtime/arch/arm64/context_arm64.cc b/runtime/arch/arm64/context_arm64.cc
--- a/runtime/arch/arm64/context_arm64.cc
+++ b/runtime/arch/arm64/context_arm64.cc
@@ -33,6 +33,6 @@ namespace art HIDDEN {
 namespace arm64 {
 
-static constexpr uint64_t gZero = 0;
+static constexpr uintptr_t gZero = 0;
 
 void Arm64Context::Reset() {
   std::fill_n(gprs_, arraysize(gprs_), nullptr);
@@ -83,6 +83,6 @@ void Arm64Context::SetFPR(uint32_t reg, uintptr_t value) {
 void Arm64Context::SmashCallerSaves() {
   // This needs to be 0 because we want a null/zero return value.
-  gprs_[X0] = const_cast<uint64_t*>(&gZero);
+  gprs_[X0] = const_cast<uintptr_t*>(&gZero);
   gprs_[X1] = nullptr;
   gprs_[X2] = nullptr;
   gprs_[X3] = nullptr;
