diff --git a/runtime/thread_list.cc b/runtime/thread_list.cc
--- a/runtime/thread_list.cc
+++ b/runtime/thread_list.cc
@@ -749,7 +749,7 @@ static bool WaitOnceForSuspendBarrier(AtomicInteger* barrier,
                                       int32_t cur_val,
                                       uint64_t timeout_ns) {
   // In the normal case, aim for a couple of hundred milliseconds.
-  static constexpr unsigned kInnerIters =
+  const unsigned kInnerIters =
       kShortSuspendTimeouts ? 1'000 : (timeout_ns / 1000) / kSuspendBarrierIters;
   DCHECK_GE(kInnerIters, 1'000u);
   for (int i = 0; i < kInnerIters; ++i) {
@@ -765,9 +765,7 @@ static bool WaitOnceForSuspendBarrier(AtomicInteger* barrier,
 
 std::optional<std::string> ThreadList::WaitForSuspendBarrier(AtomicInteger* barrier,
                                                              pid_t t,
                                                              int attempt_of_4) {
-#if ART_USE_FUTEXES
   const uint64_t start_time = NanoTime();
-#endif
   uint64_t timeout_ns =
       attempt_of_4 == 0 ? thread_suspend_timeout_ns_ : thread_suspend_timeout_ns_ / 4;
