diff --git a/runtime/arch/arm64/quick_entrypoints_arm64.S b/runtime/arch/arm64/quick_entrypoints_arm64.S
--- a/runtime/arch/arm64/quick_entrypoints_arm64.S
+++ b/runtime/arch/arm64/quick_entrypoints_arm64.S
@@ -2417,14 +2417,26 @@
     // move some code after the array switch cases and return switch cases.
     .balign 256
+#if defined(__APPLE__)
+    .private_extern _art_quick_read_barrier_mark_introspection_arrays
+    .global _art_quick_read_barrier_mark_introspection_arrays
+_art_quick_read_barrier_mark_introspection_arrays:
+#else
     .hidden art_quick_read_barrier_mark_introspection_arrays
     .global art_quick_read_barrier_mark_introspection_arrays
 art_quick_read_barrier_mark_introspection_arrays:
+#endif
     FOR_XREGISTERS INTROSPECTION_ARRAY_LOAD, BRK0_BRK0
 .Lmark_introspection_return_switch:
     FOR_WREGISTERS MOV_WIP0_TO_WREG_AND_BL_LR, BRK0_BRK0
+#if defined(__APPLE__)
+    .private_extern _art_quick_read_barrier_mark_introspection_gc_roots
+    .global _art_quick_read_barrier_mark_introspection_gc_roots
+_art_quick_read_barrier_mark_introspection_gc_roots:
+#else
     .hidden art_quick_read_barrier_mark_introspection_gc_roots
     .global art_quick_read_barrier_mark_introspection_gc_roots
 art_quick_read_barrier_mark_introspection_gc_roots:
+#endif
     READ_BARRIER_MARK_INTROSPECTION_SLOW_PATH BAKER_MARK_INTROSPECTION_GC_ROOT_LDR_OFFSET
 END art_quick_read_barrier_mark_introspection
