diff --git a/runtime/oat/oat_file.cc b/runtime/oat/oat_file.cc
--- a/runtime/oat/oat_file.cc
+++ b/runtime/oat/oat_file.cc
@@ -354,7 +354,12 @@ OatFileBase* OatFileBase::OpenOatFileFromSdm(const std::string& sdm_filename,
   if (sdc_reader == nullptr) {
     return nullptr;
   }
-  if (sdc_reader->GetSdmTimestampNs() != TimeSpecToNs(sdm_st.st_mtim)) {
+#if defined(__APPLE__)
+  const timespec& sdm_mtime = sdm_st.st_mtimespec;
+#else
+  const timespec& sdm_mtime = sdm_st.st_mtim;
+#endif
+  if (sdc_reader->GetSdmTimestampNs() != TimeSpecToNs(sdm_mtime)) {
     // The sdm file had been replaced after the sdc file was created.
     *error_msg = ART_FORMAT("Obsolete sdc file '{}'", sdc_filename);
     return nullptr;
