diff --git a/runtime/exec_utils.cc b/runtime/exec_utils.cc
--- a/runtime/exec_utils.cc
+++ b/runtime/exec_utils.cc
@@ -17,6 +17,7 @@
 #include "exec_utils.h"
 
 #include <poll.h>
+#include <signal.h>
 #include <sys/types.h>
 #include <sys/wait.h>
 #include <unistd.h>
@@ -41,6 +42,8 @@
 #include "android-base/unique_fd.h"
 #include "base/macros.h"
+#if !defined(__APPLE__)
 #include "base/pidfd.h"
+#endif
 #include "base/utils.h"
 #include "runtime.h"
 
@@ -326,7 +329,13 @@ bool ExecUtils::Exec(const std::vector<std::string>& arg_vector, std::string* error_msg) const {
   return true;
 }
 
-unique_fd ExecUtils::PidfdOpen(pid_t pid) const { return art::PidfdOpen(pid, /*flags=*/0); }
+unique_fd ExecUtils::PidfdOpen(pid_t pid) const {
+#if defined(__APPLE__)
+  return unique_fd(-1);
+#else
+  return art::PidfdOpen(pid, /*flags=*/0);
+#endif
+}
 
 std::string ExecUtils::GetProcStat(pid_t pid) const {
   std::string stat_content;
