diff --git a/runtime/arch/arm64/instruction_set_features_arm64.cc b/runtime/arch/arm64/instruction_set_features_arm64.cc
--- a/runtime/arch/arm64/instruction_set_features_arm64.cc
+++ b/runtime/arch/arm64/instruction_set_features_arm64.cc
@@ -31,12 +31,14 @@
 #include "base/array_ref.h"
 #include "base/stl_util.h"
 
+#if !defined(__APPLE__)
 #include <cpu_features_macros.h>
 
 #ifdef CPU_FEATURES_ARCH_AARCH64
 // This header can only be included on aarch64 targets,
 // as determined by cpu_features own define.
 #include <cpuinfo_aarch64.h>
 #endif
+#endif
 
 namespace art HIDDEN {
