diff --git a/runtime/arch/arm64/native_entrypoints_arm64.S b/runtime/arch/arm64/native_entrypoints_arm64.S
--- a/runtime/arch/arm64/native_entrypoints_arm64.S
+++ b/runtime/arch/arm64/native_entrypoints_arm64.S
@@ -69,10 +69,18 @@ ENTRY art_jni_dlsym_lookup_stub
 #endif
     tst   xIP0, xIP1
     b.ne  .Llookup_stub_fast_or_critical_native
+#if defined(__APPLE__)
+    bl    _artFindNativeMethod
+#else
     bl    artFindNativeMethod
+#endif
     b     .Llookup_stub_continue
     .Llookup_stub_fast_or_critical_native:
+#if defined(__APPLE__)
+    bl    _artFindNativeMethodRunnable
+#else
     bl    artFindNativeMethodRunnable
+#endif
 .Llookup_stub_continue:
     mov   x17, x0    // store result in scratch reg.
 
diff --git a/runtime/arch/arm64/quick_entrypoints_arm64.S b/runtime/arch/arm64/quick_entrypoints_arm64.S
--- a/runtime/arch/arm64/quick_entrypoints_arm64.S
+++ b/runtime/arch/arm64/quick_entrypoints_arm64.S
@@ -2600,7 +2600,11 @@ ENTRY art_quick_record_long_running_entry_trace_event
     ldr x0, [sp, #FRAME_SIZE_SAVE_EVERYTHING] // pass ArtMethod*
     mov x1, xSELF                             // pass Thread::Current
     mov x2, 1                                 // set to true for entry events
+#if defined(__APPLE__)
+    bl  _artRecordLongRunningMethodTraceEvent // (ArtMethod*, Thread*, is_entry)
+#else
     bl  artRecordLongRunningMethodTraceEvent // (ArtMethod*, Thread*, is_entry)
+#endif
     RESTORE_SAVE_EVERYTHING_FRAME             // Note: will restore xSELF
     REFRESH_MARKING_REGISTER
     ret
@@ -2653,7 +2657,11 @@ ENTRY art_quick_record_long_running_exit_trace_event
     ldr xzr, [sp, #FRAME_SIZE_SAVE_EVERYTHING] // pass nullptr. ArtMethod* isn't required for exits
     mov x1, xSELF                              // pass Thread::Current
     mov x2, 0                                  // set to false for exit events
+#if defined(__APPLE__)
+    bl  _artRecordLongRunningMethodTraceEvent   // (ArtMethod*, Thread*, is_entry)
+#else
     bl  artRecordLongRunningMethodTraceEvent   // (ArtMethod*, Thread*, is_entry)
+#endif
     RESTORE_SAVE_EVERYTHING_FRAME              // Note: will restore xSELF
     REFRESH_MARKING_REGISTER
     ret
