--- a/runtime/runtime_common.cc
+++ b/runtime/runtime_common.cc
@@ -19,5 +19,9 @@
 #include <signal.h>
 
+#if defined(__APPLE__) && defined(__aarch64__)
+#include <mach/arm/thread_status.h>
+#endif
+
 #include <cinttypes>
 #include <iostream>
 #include <sstream>
@@ -148,7 +152,22 @@
 };
 
 void UContext::Dump(std::ostream& os) const {
-#if defined(__APPLE__) && defined(__i386__)
+#if defined(__APPLE__) && defined(__aarch64__)
+  for (size_t i = 0; i <= 28; ++i) {
+    std::string reg_name = "x" + std::to_string(i);
+    DumpRegister64(os, reg_name.c_str(), context->__ss.__x[i]);
+    if (i % 4 == 3) os << '\n';
+  }
+  os << '\n';
+  DumpRegister64(os, "fp", arm_thread_state64_get_fp(context->__ss));
+  DumpRegister64(os, "lr", arm_thread_state64_get_lr(context->__ss));
+  DumpRegister64(os, "sp", arm_thread_state64_get_sp(context->__ss));
+  DumpRegister64(os, "pc", arm_thread_state64_get_pc(context->__ss));
+  os << '\n';
+  DumpRegister64(os, "pstate", context->__ss.__cpsr);
+  DumpArmStatusRegister(os, context->__ss.__cpsr);
+  os << '\n';
+#elif defined(__APPLE__) && defined(__i386__)
   DumpRegister32(os, "eax", context->__ss.__eax);
   DumpRegister32(os, "ebx", context->__ss.__ebx);
   DumpRegister32(os, "ecx", context->__ss.__ecx);
@@ -407,9 +426,21 @@
                              GetSignalCodeName(signal_number, info->si_code))
            << (has_address ? StringPrintf(" fault addr %p", info->si_addr) : "") << std::endl
            << "OS: " << Dumpable<OsInfo>(os_info) << std::endl
+           << "Process: " << getpid() << std::endl
            << "Cmdline: " << cmd_line << std::endl
            << "Thread: " << tid << " \"" << thread_name << "\"" << std::endl
-           << "Registers:\n" << Dumpable<UContext>(thread_context) << std::endl
-           << "Backtrace:\n" << Dumpable<Backtrace>(thread_backtrace) << std::endl;
+           << "Registers:\n" << Dumpable<UContext>(thread_context) << std::endl;
+#if defined(__APPLE__) && defined(__aarch64__)
+    uintptr_t pc = arm_thread_state64_get_pc(
+        reinterpret_cast<ucontext_t*>(raw_context)->uc_mcontext->__ss);
+    const uint32_t* instructions = reinterpret_cast<const uint32_t*>(pc) - 16;
+    stream << "Code around pc:" << std::hex;
+    for (size_t i = 0; i != 24; ++i) {
+      if (i % 8 == 0) stream << '\n' << reinterpret_cast<const void*>(instructions + i) << ":";
+      stream << ' ' << instructions[i];
+    }
+    stream << std::dec << std::endl;
+#endif
+    stream << "Backtrace:\n" << Dumpable<Backtrace>(thread_backtrace) << std::endl;
     stream << std::flush;
   };
