diff --git a/runtime/mirror/object_reference.h b/runtime/mirror/object_reference.h
--- a/runtime/mirror/object_reference.h
+++ b/runtime/mirror/object_reference.h
@@ -97,12 +97,24 @@ class PtrCompression {
  public:
   // Compress reference to its bit representation.
   static uint32_t Compress(MirrorType* mirror_ptr) {
-    uint32_t as_bits = reinterpret_cast32<uint32_t>(mirror_ptr);
+    uintptr_t address = reinterpret_cast<uintptr_t>(mirror_ptr);
+    uint32_t as_bits;
+    if (address == 0u) {
+      as_bits = 0u;
+    } else {
+      DCHECK_GE(address, kArtCompressedReferenceBase);
+      DCHECK_LT(address, kArtCompressedReferenceLimit);
+      as_bits = static_cast<uint32_t>(address - kArtCompressedReferenceBase);
+      DCHECK_NE(as_bits, 0u);
+    }
     return kPoisonReferences ? -as_bits : as_bits;
   }
 
   // Uncompress an encoded reference from its bit representation.
   static MirrorType* Decompress(uint32_t ref) {
     uint32_t as_bits = kPoisonReferences ? -ref : ref;
-    return reinterpret_cast32<MirrorType*>(as_bits);
+    if (as_bits == 0u) {
+      return nullptr;
+    }
+    return reinterpret_cast<MirrorType*>(kArtCompressedReferenceBase + as_bits);
   }
