diff --git a/runtime/entrypoints/quick/quick_trampoline_entrypoints.cc b/runtime/entrypoints/quick/quick_trampoline_entrypoints.cc
--- a/runtime/entrypoints/quick/quick_trampoline_entrypoints.cc
+++ b/runtime/entrypoints/quick/quick_trampoline_entrypoints.cc
@@ -2770,5 +2770,9 @@
     if (is_ref) {
       // Restore the return value if it's a reference since it might have moved.
+#if defined(__APPLE__)
+      *gpr_result = mirror::PtrCompression<false, mirror::Object>::Compress(res.Get());
+#else
       *reinterpret_cast<mirror::Object**>(gpr_result) = res.Get();
+#endif
       return_value.SetL(res.Get());
     }
diff --git a/runtime/instrumentation.cc b/runtime/instrumentation.cc
--- a/runtime/instrumentation.cc
+++ b/runtime/instrumentation.cc
@@ -1663,5 +1663,10 @@
   } else if (return_shorty == 'F' || return_shorty == 'D') {
     return_value.SetJ(*fpr_result);
+#if defined(__APPLE__)
+  } else if (*is_ref) {
+    return_value.SetL(mirror::PtrCompression<false, mirror::Object>::Decompress(
+        static_cast<uint32_t>(*gpr_result)));
+#endif
   } else {
     return_value.SetJ(*gpr_result);
   }
