diff --git a/compiler/optimizing/jit_patches_arm64.h b/compiler/optimizing/jit_patches_arm64.h
--- a/compiler/optimizing/jit_patches_arm64.h
+++ b/compiler/optimizing/jit_patches_arm64.h
@@ -69,1 +69,1 @@
-                                          vixl::aarch64::Literal<uint32_t>*,
+                                          vixl::aarch64::Literal<uint64_t>*,
@@ -72,1 +72,1 @@
-                                        vixl::aarch64::Literal<uint32_t>*,
+                                        vixl::aarch64::Literal<uint64_t>*,
@@ -75,1 +75,1 @@
-                                         vixl::aarch64::Literal<uint32_t>*,
+                                         vixl::aarch64::Literal<uint64_t>*,
@@ -81,1 +81,1 @@
-  vixl::aarch64::Literal<uint32_t>* DeduplicateJitStringLiteral(
+  vixl::aarch64::Literal<uint64_t>* DeduplicateJitStringLiteral(
@@ -86,1 +86,1 @@
-  vixl::aarch64::Literal<uint32_t>* DeduplicateJitClassLiteral(
+  vixl::aarch64::Literal<uint64_t>* DeduplicateJitClassLiteral(
@@ -91,1 +91,1 @@
-  vixl::aarch64::Literal<uint32_t>* DeduplicateJitMethodTypeLiteral(
+  vixl::aarch64::Literal<uint64_t>* DeduplicateJitMethodTypeLiteral(
diff --git a/compiler/optimizing/jit_patches_arm64.cc b/compiler/optimizing/jit_patches_arm64.cc
--- a/compiler/optimizing/jit_patches_arm64.cc
+++ b/compiler/optimizing/jit_patches_arm64.cc
@@ -45,1 +45,1 @@
-                            vixl::aarch64::Literal<uint32_t>* literal,
+                            vixl::aarch64::Literal<uint64_t>* literal,
@@ -51,1 +51,1 @@
-  reinterpret_cast<uint32_t*>(data)[0] = dchecked_integral_cast<uint32_t>(address);
+  reinterpret_cast<uint64_t*>(data)[0] = dchecked_integral_cast<uint64_t>(address);
@@ -60,1 +60,1 @@
-    vixl::aarch64::Literal<uint32_t>* table_entry_literal = entry.second;
+    vixl::aarch64::Literal<uint64_t>* table_entry_literal = entry.second;
@@ -66,1 +66,1 @@
-    vixl::aarch64::Literal<uint32_t>* table_entry_literal = entry.second;
+    vixl::aarch64::Literal<uint64_t>* table_entry_literal = entry.second;
@@ -72,1 +72,1 @@
-    vixl::aarch64::Literal<uint32_t>* table_entry_literal = entry.second;
+    vixl::aarch64::Literal<uint64_t>* table_entry_literal = entry.second;
@@ -83,1 +83,1 @@
-vixl::aarch64::Literal<uint32_t>* JitPatchesARM64::DeduplicateJitStringLiteral(
+vixl::aarch64::Literal<uint64_t>* JitPatchesARM64::DeduplicateJitStringLiteral(
@@ -92,1 +92,1 @@
-        return GetVIXLAssembler()->CreateLiteralDestroyedWithPool<uint32_t>(/* value= */ 0u);
+        return GetVIXLAssembler()->CreateLiteralDestroyedWithPool<uint64_t>(/* value= */ 0u);
@@ -96,1 +96,1 @@
-vixl::aarch64::Literal<uint32_t>* JitPatchesARM64::DeduplicateJitClassLiteral(
+vixl::aarch64::Literal<uint64_t>* JitPatchesARM64::DeduplicateJitClassLiteral(
@@ -105,1 +105,1 @@
-        return GetVIXLAssembler()->CreateLiteralDestroyedWithPool<uint32_t>(/* value= */ 0u);
+        return GetVIXLAssembler()->CreateLiteralDestroyedWithPool<uint64_t>(/* value= */ 0u);
@@ -109,1 +109,1 @@
-vixl::aarch64::Literal<uint32_t>* JitPatchesARM64::DeduplicateJitMethodTypeLiteral(
+vixl::aarch64::Literal<uint64_t>* JitPatchesARM64::DeduplicateJitMethodTypeLiteral(
@@ -118,1 +118,1 @@
-        return GetVIXLAssembler()->CreateLiteralDestroyedWithPool<uint32_t>(/* value= */ 0u);
+        return GetVIXLAssembler()->CreateLiteralDestroyedWithPool<uint64_t>(/* value= */ 0u);
diff --git a/compiler/optimizing/code_generator_arm64.h b/compiler/optimizing/code_generator_arm64.h
--- a/compiler/optimizing/code_generator_arm64.h
+++ b/compiler/optimizing/code_generator_arm64.h
@@ -933,1 +933,1 @@
-  vixl::aarch64::Literal<uint32_t>* DeduplicateJitStringLiteral(const DexFile& dex_file,
+  vixl::aarch64::Literal<uint64_t>* DeduplicateJitStringLiteral(const DexFile& dex_file,
@@ -939,1 +939,1 @@
-  vixl::aarch64::Literal<uint32_t>* DeduplicateJitClassLiteral(const DexFile& dex_file,
+  vixl::aarch64::Literal<uint64_t>* DeduplicateJitClassLiteral(const DexFile& dex_file,
@@ -945,1 +945,1 @@
-  vixl::aarch64::Literal<uint32_t>* DeduplicateJitMethodTypeLiteral(
+  vixl::aarch64::Literal<uint64_t>* DeduplicateJitMethodTypeLiteral(
diff --git a/compiler/optimizing/code_generator_arm64.cc b/compiler/optimizing/code_generator_arm64.cc
--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -5828,1 +5828,1 @@
-      __ Ldr(out, codegen_->DeduplicateJitClassLiteral(cls->GetDexFile(),
+      __ Ldr(out.X(), codegen_->DeduplicateJitClassLiteral(cls->GetDexFile(),
@@ -5922,1 +5922,1 @@
-      __ Ldr(out, codegen_->DeduplicateJitMethodTypeLiteral(load->GetDexFile(),
+      __ Ldr(out.X(), codegen_->DeduplicateJitMethodTypeLiteral(load->GetDexFile(),
@@ -6059,1 +6059,1 @@
-      __ Ldr(out, codegen_->DeduplicateJitStringLiteral(load->GetDexFile(),
+      __ Ldr(out.X(), codegen_->DeduplicateJitStringLiteral(load->GetDexFile(),

