diff --git a/compiler/optimizing/code_generator_arm64.cc b/compiler/optimizing/code_generator_arm64.cc
--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -5819,7 +5819,8 @@
     case HLoadClass::LoadKind::kJitBootImageAddress: {
       DCHECK_EQ(read_barrier_option, kWithoutReadBarrier);
-      uint32_t address = reinterpret_cast32<uint32_t>(cls->GetClass().Get());
+      uint32_t address = mirror::CompressedReference<mirror::Object>::FromMirrorPtr(
+          cls->GetClass().Get()).AsVRegValue();
       DCHECK_NE(address, 0u);
       __ Ldr(out.W(), codegen_->DeduplicateBootImageAddressLiteral(address));
       break;
     }
@@ -6051,7 +6052,8 @@
     case HLoadString::LoadKind::kJitBootImageAddress: {
-      uint32_t address = reinterpret_cast32<uint32_t>(load->GetString().Get());
+      uint32_t address = mirror::CompressedReference<mirror::Object>::FromMirrorPtr(
+          load->GetString().Get()).AsVRegValue();
       DCHECK_NE(address, 0u);
       __ Ldr(out.W(), codegen_->DeduplicateBootImageAddressLiteral(address));
       return;
     }
     case HLoadString::LoadKind::kJitTableAddress: {
