--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -6196,7 +6196,9 @@
 void InstructionCodeGeneratorARM64::VisitNewArray(HNewArray* instruction) {
   // Note: if heap poisoning is enabled, the entry point takes care of poisoning the reference.
   QuickEntrypointEnum entrypoint = CodeGenerator::GetArrayAllocationEntrypoint(instruction);
+  ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), x0, x0);
   codegen_->InvokeRuntime(entrypoint, instruction);
+  ReferenceCodegenARM64<>::Encode(GetVIXLAssembler(), x0, x0);
   CheckEntrypointTypes<kQuickAllocArrayResolved, void*, mirror::Class*, int32_t>();
   codegen_->MaybeGenerateMarkingRegisterCheck(/* code= */ __LINE__);
 }
@@ -6210,7 +6212,9 @@
 }
 
 void InstructionCodeGeneratorARM64::VisitNewInstance(HNewInstance* instruction) {
+  ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), x0, x0);
   codegen_->InvokeRuntime(instruction->GetEntrypoint(), instruction);
+  ReferenceCodegenARM64<>::Encode(GetVIXLAssembler(), x0, x0);
   CheckEntrypointTypes<kQuickAllocObjectWithChecks, void*, mirror::Class*>();
   codegen_->MaybeGenerateMarkingRegisterCheck(/* code= */ __LINE__);
 }
