--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -482,3 +482,5 @@
                                DataType::Type::kReference);
     if (instruction_->IsInstanceOf()) {
+      ReferenceCodegenARM64<>::DecodeNullable(arm64_codegen->GetVIXLAssembler(), x0, w0);
+      ReferenceCodegenARM64<>::DecodeNullable(arm64_codegen->GetVIXLAssembler(), x1, w1);
       arm64_codegen->InvokeRuntime(kQuickInstanceofNonTrivial, instruction_, this);
@@ -490,3 +492,5 @@
     } else {
       DCHECK(instruction_->IsCheckCast());
+      ReferenceCodegenARM64<>::DecodeNullable(arm64_codegen->GetVIXLAssembler(), x0, w0);
+      ReferenceCodegenARM64<>::DecodeNullable(arm64_codegen->GetVIXLAssembler(), x1, w1);
       arm64_codegen->InvokeRuntime(kQuickCheckInstanceOf, instruction_, this);
@@ -2213,14 +2213,17 @@
   uint32_t mask = check->GetBitstringMask();
   DCHECK(IsPowerOfTwo(mask + 1));
   size_t mask_bits = WhichPowerOf2(mask + 1);
+  UseScratchRegisterScope temps(GetVIXLAssembler());
+  Register base = temps.AcquireX();
+  ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), base, temp.W());
 
   if (mask_bits == 16u) {
     // Load only the bitstring part of the status word.
-    __ Ldrh(temp, HeapOperand(temp, mirror::Class::StatusOffset()));
+    __ Ldrh(temp, MemOperand(base, mirror::Class::StatusOffset().Int32Value()));
   } else {
     // /* uint32_t */ temp = temp->status_
-    __ Ldr(temp, HeapOperand(temp, mirror::Class::StatusOffset()));
+    __ Ldr(temp, MemOperand(base, mirror::Class::StatusOffset().Int32Value()));
     // Extract the bitstring bits.
     __ Ubfx(temp, temp, 0, mask_bits);
   }
   // Compare the bitstring bits to `path_to_root`.
@@ -7000,4 +7000,7 @@
     // Plain load with no read barrier.
     // /* HeapReference<Object> */ out = *(out + offset)
-    __ Ldr(out_reg, HeapOperand(out_reg, offset));
+    UseScratchRegisterScope temps(GetVIXLAssembler());
+    Register base = temps.AcquireX();
+    ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), base, out_reg.W());
+    __ Ldr(out_reg, MemOperand(base, offset));
     GetAssembler()->MaybeUnpoisonHeapReference(out_reg);
@@ -7040,4 +7043,7 @@
     // Plain load with no read barrier.
     // /* HeapReference<Object> */ out = *(obj + offset)
-    __ Ldr(out_reg, HeapOperand(obj_reg, offset));
+    UseScratchRegisterScope temps(GetVIXLAssembler());
+    Register base = temps.AcquireX();
+    ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), base, obj_reg.W());
+    __ Ldr(out_reg, MemOperand(base, offset));
     GetAssembler()->MaybeUnpoisonHeapReference(out_reg);
