--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -4550,13 +4550,16 @@
       Register temp = temps.AcquireW();
       Register temp2 = temps.AcquireW();
       // /* HeapReference<Class> */ temp = obj->klass_
-      __ Ldr(temp, HeapOperand(obj, class_offset));
+      ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), temp.X(), obj.W());
+      __ Ldr(temp, MemOperand(temp.X(), class_offset));
       GetAssembler()->MaybeUnpoisonHeapReference(temp);
       // /* HeapReference<Class> */ temp = temp->iftable_
-      __ Ldr(temp, HeapOperand(temp, iftable_offset));
+      ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), temp.X(), temp.W());
+      __ Ldr(temp, MemOperand(temp.X(), iftable_offset));
       GetAssembler()->MaybeUnpoisonHeapReference(temp);
       // Load the size of the `IfTable`. The `Class::iftable_` is never null.
-      __ Ldr(out, HeapOperand(temp, array_length_offset));
+      ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), temp.X(), temp.W());
+      __ Ldr(out, MemOperand(temp.X(), array_length_offset));
       // Loop through the `IfTable` and check if any class matches.
       vixl::aarch64::Label loop;
       __ Bind(&loop);
@@ -4564,6 +4567,6 @@
       __ Cbz(out, &done);  // If taken, the result in `out` is already 0 (false).
-      __ Ldr(temp2, HeapOperand(temp, object_array_data_offset));
+      __ Ldr(temp2, MemOperand(temp.X(), object_array_data_offset));
       GetAssembler()->MaybeUnpoisonHeapReference(temp2);
       // Go to next interface.
-      __ Add(temp, temp, 2 * kHeapReferenceSize);
+      __ Add(temp.X(), temp.X(), 2 * kHeapReferenceSize);
       __ Sub(out, out, 2);
@@ -4800,11 +4800,12 @@
       // Load the size of the `IfTable`. The `Class::iftable_` is never null.
-      __ Ldr(WRegisterFrom(maybe_temp2_loc), HeapOperand(temp.W(), array_length_offset));
+      ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), temp.X(), temp.W());
+      __ Ldr(WRegisterFrom(maybe_temp2_loc), MemOperand(temp.X(), array_length_offset));
       // Loop through the iftable and check if any class matches.
       vixl::aarch64::Label start_loop;
       __ Bind(&start_loop);
       __ Cbz(WRegisterFrom(maybe_temp2_loc), type_check_slow_path->GetEntryLabel());
-      __ Ldr(WRegisterFrom(maybe_temp3_loc), HeapOperand(temp.W(), object_array_data_offset));
+      __ Ldr(WRegisterFrom(maybe_temp3_loc), MemOperand(temp.X(), object_array_data_offset));
       GetAssembler()->MaybeUnpoisonHeapReference(WRegisterFrom(maybe_temp3_loc));
       // Go to next interface.
-      __ Add(temp, temp, 2 * kHeapReferenceSize);
+      __ Add(temp.X(), temp.X(), 2 * kHeapReferenceSize);
       __ Sub(WRegisterFrom(maybe_temp2_loc), WRegisterFrom(maybe_temp2_loc), 2);
