--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -4500,4 +4500,5 @@
       // If `out` is null, we use it for the result, and jump to `done`.
       __ Cbz(out, &done);
-      __ Ldrh(out, HeapOperand(out, primitive_offset));
+      ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), out.X(), out.W());
+      __ Ldrh(out, MemOperand(out.X(), primitive_offset));
       static_assert(Primitive::kPrimNot == 0, "Expected 0 for kPrimNot");
@@ -4790,4 +4791,5 @@
       // primitive type.
-      __ Ldrh(temp, HeapOperand(temp, primitive_offset));
+      ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), temp.X(), temp.W());
+      __ Ldrh(temp, MemOperand(temp.X(), primitive_offset));
       static_assert(Primitive::kPrimNot == 0, "Expected 0 for kPrimNot");
       __ Cbnz(temp, type_check_slow_path->GetEntryLabel());
