--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -4987,13 +4987,15 @@
   if (receiver.IsStackSlot()) {
     __ Ldr(temp.W(), StackOperandFrom(receiver));
+    ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), temp, temp);
     {
       EmissionCheckScope guard(GetVIXLAssembler(), kMaxMacroInstructionSizeInBytes);
       // /* HeapReference<Class> */ temp = temp->klass_
-      __ Ldr(temp.W(), HeapOperand(temp.W(), class_offset));
+      __ Ldr(temp.W(), MemOperand(temp, class_offset.SizeValue()));
       codegen_->MaybeRecordImplicitNullCheck(invoke);
     }
   } else {
+    ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), temp, WRegisterFrom(receiver));
     EmissionCheckScope guard(GetVIXLAssembler(), kMaxMacroInstructionSizeInBytes);
     // /* HeapReference<Class> */ temp = receiver->klass_
-    __ Ldr(temp.W(), HeapOperandFrom(receiver, class_offset));
+    __ Ldr(temp.W(), MemOperand(temp, class_offset.SizeValue()));
     codegen_->MaybeRecordImplicitNullCheck(invoke);
@@ -5035,2 +5037,3 @@
+  ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), temp, temp);
   __ Ldr(temp,
       MemOperand(temp, mirror::Class::ImtPtrOffset(kArm64PointerSize).Uint32Value()));
