--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -6650,8 +6650,12 @@
 void InstructionCodeGeneratorARM64::VisitUnresolvedInstanceFieldGet(
     HUnresolvedInstanceFieldGet* instruction) {
   FieldAccessCallingConventionARM64 calling_convention;
+  ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), x1, w1);
   codegen_->GenerateUnresolvedFieldAccess(instruction,
                                           instruction->GetFieldType(),
                                           instruction->GetFieldIndex(),
                                           calling_convention);
+  if (instruction->GetFieldType() == DataType::Type::kReference) {
+    ReferenceCodegenARM64<>::Encode(GetVIXLAssembler(), x0, x0);
+  }
 }
@@ -6666,8 +6670,12 @@
 void InstructionCodeGeneratorARM64::VisitUnresolvedInstanceFieldSet(
     HUnresolvedInstanceFieldSet* instruction) {
   FieldAccessCallingConventionARM64 calling_convention;
+  ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), x1, w1);
+  if (instruction->GetFieldType() == DataType::Type::kReference) {
+    ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), x2, w2);
+  }
   codegen_->GenerateUnresolvedFieldAccess(instruction,
                                           instruction->GetFieldType(),
                                           instruction->GetFieldIndex(),
                                           calling_convention);
 }
@@ -6682,8 +6682,11 @@
 void InstructionCodeGeneratorARM64::VisitUnresolvedStaticFieldGet(
     HUnresolvedStaticFieldGet* instruction) {
   FieldAccessCallingConventionARM64 calling_convention;
   codegen_->GenerateUnresolvedFieldAccess(instruction,
                                           instruction->GetFieldType(),
                                           instruction->GetFieldIndex(),
                                           calling_convention);
+  if (instruction->GetFieldType() == DataType::Type::kReference) {
+    ReferenceCodegenARM64<>::Encode(GetVIXLAssembler(), x0, x0);
+  }
 }
@@ -6698,8 +6701,11 @@
 void InstructionCodeGeneratorARM64::VisitUnresolvedStaticFieldSet(
     HUnresolvedStaticFieldSet* instruction) {
   FieldAccessCallingConventionARM64 calling_convention;
+  if (instruction->GetFieldType() == DataType::Type::kReference) {
+    ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), x1, w1);
+  }
   codegen_->GenerateUnresolvedFieldAccess(instruction,
                                           instruction->GetFieldType(),
                                           instruction->GetFieldIndex(),
                                           calling_convention);
 }
