diff --git a/compiler/optimizing/fast_compiler_arm64.cc b/compiler/optimizing/fast_compiler_arm64.cc
--- a/compiler/optimizing/fast_compiler_arm64.cc
+++ b/compiler/optimizing/fast_compiler_arm64.cc
@@ -983,7 +983,7 @@ bool FastCompilerARM64::BuildLoadString(uint32_t vreg,
 
   Handle<mirror::String> h_str = handles_->NewHandle(str);
   Register dst = RegisterFrom(loc, DataType::Type::kReference);
-  __ Ldr(dst.W(), jit_patches_.DeduplicateJitStringLiteral(GetDexFile(),
+  __ Ldr(dst.X(), jit_patches_.DeduplicateJitStringLiteral(GetDexFile(),
                                                            string_index,
                                                            h_str,
                                                            code_generation_data_.get()));
@@ -1019,7 +1019,7 @@ bool FastCompilerARM64::BuildNewInstance(uint32_t vreg,
   InvokeRuntimeCallingConvention calling_convention;
   Register cls_reg = calling_convention.GetRegisterAt(0);
   Handle<mirror::Class> h_klass = handles_->NewHandle(klass);
-  __ Ldr(cls_reg.W(), jit_patches_.DeduplicateJitClassLiteral(GetDexFile(),
+  __ Ldr(cls_reg.X(), jit_patches_.DeduplicateJitClassLiteral(GetDexFile(),
                                                               type_index,
                                                               h_klass ,
                                                               code_generation_data_.get()));
@@ -1073,7 +1073,7 @@ bool FastCompilerARM64::BuildCheckCast(uint32_t vreg, dex::TypeIndex type_index, uint32_t dex_pc) {
 
   vixl::aarch64::Label exit, read_barrier_exit;
   __ Cbz(obj, &exit);
-  __ Ldr(cls.W(), jit_patches_.DeduplicateJitClassLiteral(GetDexFile(),
+  __ Ldr(cls.X(), jit_patches_.DeduplicateJitClassLiteral(GetDexFile(),
                                                           type_index,
                                                           h_klass ,
                                                           code_generation_data_.get()));
