--- a/compiler/optimizing/optimizing_compiler.cc
+++ b/compiler/optimizing/optimizing_compiler.cc
@@ -1017,12 +1017,6 @@
         }
       }
-      if (gUseReadBarrier && (instruction->IsUnresolvedInstanceFieldGet() ||
-          instruction->IsUnresolvedInstanceFieldSet() ||
-          instruction->IsUnresolvedStaticFieldGet() ||
-          instruction->IsUnresolvedStaticFieldSet())) {
-        VLOG(jit) << "Darwin graph reject " << instruction->DebugName() << " in " << graph->PrettyMethod(); return nullptr;
-      }
       if (!instruction->IsInvoke()) continue;
       if ((instruction->IsInvokeUnresolved() || instruction->IsInvokePolymorphic()) && !gUseReadBarrier) continue;
       if (instruction->IsInvokeVirtual() || instruction->IsInvokeInterface()) {
