--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -7662,7 +7662,13 @@
       UseScratchRegisterScope temps(assembler.GetVIXLAssembler());
       temps.Exclude(ip0, ip1);
       vixl::aarch64::Label return_label, not_marked, forwarding_address;
       __ Cbz(root_reg, &return_label);
-      MemOperand lock_word(root_reg.X(), mirror::Object::MonitorOffset().Int32Value());
+      Register lock_word_base = root_reg.X();
+      if (ReferenceCodegenARM64<>::kNeedsHeapBase) {
+        ReferenceCodegenARM64<>::DecodeNonNull(
+            assembler.GetVIXLAssembler(), ip1, root_reg);
+        lock_word_base = ip1;
+      }
+      MemOperand lock_word(lock_word_base, mirror::Object::MonitorOffset().Int32Value());
       __ Ldr(ip0.W(), lock_word);
       __ Tbz(ip0.W(), LockWord::kMarkBitStateShift, &not_marked);
       __ Bind(&return_label);
