diff --git a/compiler/optimizing/optimizing_compiler.cc b/compiler/optimizing/optimizing_compiler.cc
--- a/compiler/optimizing/optimizing_compiler.cc
+++ b/compiler/optimizing/optimizing_compiler.cc
@@ -1016,7 +1016,7 @@
           VLOG(jit) << "Darwin graph reject " << instruction->DebugName() << " in " << graph->PrettyMethod(); return nullptr;
         }
       }
       if (!instruction->IsInvoke()) continue;
-      if ((instruction->IsInvokeUnresolved() || instruction->IsInvokePolymorphic()) && !gUseReadBarrier) continue;
+      if (instruction->IsInvokeUnresolved() || instruction->IsInvokePolymorphic()) continue;
       if (instruction->IsInvokeVirtual() || instruction->IsInvokeInterface()) {
