diff --git a/compiler/optimizing/optimizing_compiler.cc b/compiler/optimizing/optimizing_compiler.cc
--- a/compiler/optimizing/optimizing_compiler.cc
+++ b/compiler/optimizing/optimizing_compiler.cc
@@ -1022,8 +1022,8 @@
       if (instruction->IsInvokeVirtual() || instruction->IsInvokeInterface()) {
         auto* invoke = instruction->AsInvoke();
         auto* resolved = invoke->GetResolvedMethod();
-        if (resolved == nullptr ||
-            resolved->IsStatic() || invoke->IsIntrinsic()) return nullptr;
+        if (invoke->IsIntrinsic()) continue;
+        if (resolved == nullptr || resolved->IsStatic()) return nullptr;
         continue;
       }
 
@@ -1031,6 +1031,6 @@
       HInvokeStaticOrDirect* invoke = instruction->AsInvokeStaticOrDirect();
       ArtMethod* resolved = invoke->GetResolvedMethod();
-      if (resolved == nullptr || invoke->IsIntrinsic() ||
-          invoke->IsStatic() != resolved->IsStatic() ||
+      if (invoke->IsIntrinsic()) continue;
+      if (resolved == nullptr || invoke->IsStatic() != resolved->IsStatic() ||
           invoke->GetMethodLoadKind() != MethodLoadKind::kJitDirectAddress ||
           invoke->GetCodePtrLocation() != CodePtrLocation::kCallArtMethod) {
