diff --git a/compiler/optimizing/intrinsics_arm64.cc b/compiler/optimizing/intrinsics_arm64.cc
--- a/compiler/optimizing/intrinsics_arm64.cc
+++ b/compiler/optimizing/intrinsics_arm64.cc
@@ -1012,6 +1012,7 @@ static void CreateUnsafePutLocations(ArenaAllocator* allocator, HInvoke* invoke)
   } else {
     locations->SetInAt(kValueIndex, Location::RequiresRegister());
   }
+  locations->AddTemp(FixedTempLocation());
 }
 
 static void CreateUnsafePutAbsoluteLocations(ArenaAllocator* allocator, HInvoke* invoke) {
@@ -1107,11 +1108,13 @@ static void GenUnsafePut(HInvoke* invoke,
   Location offset = locations->InAt(kOffsetIndex);      // Long offset.
   CPURegister value = InputCPURegisterOrZeroRegAt(invoke, kValueIndex);
   CPURegister source = value;
+  Register address = XRegisterFrom(locations->GetTemp(0));
+  ReferenceCodegenARM64<>::DecodeNullable(masm, address, base);
   MemOperand mem_op;
   if (offset.IsConstant()) {
-    mem_op = MemOperand(base.X(), Int64FromLocation(offset));
+    mem_op = MemOperand(address, Int64FromLocation(offset));
   } else {
-    mem_op = MemOperand(base.X(), XRegisterFrom(offset));
+    mem_op = MemOperand(address, XRegisterFrom(offset));
   }
 
   {
@@ -1575,7 +1579,8 @@ class ReadBarrierCasSlowPathARM64 : public SlowPathCodeARM64 {
     Register store_result = store_result_.IsValid() ? store_result_ : temps.AcquireW();
 
     // Recalculate the `tmp_ptr` from main path clobbered by the read barrier above.
-    __ Add(tmp_ptr, base_.X(), Operand(offset_));
+    ReferenceCodegenARM64<>::DecodeNullable(masm, tmp_ptr, base_);
+    __ Add(tmp_ptr, tmp_ptr, Operand(offset_));
 
     vixl::aarch64::Label mark_old_value;
     GenerateCompareAndSet(arm64_codegen,
@@ -1685,7 +1690,8 @@ static void GenUnsafeCas(HInvoke* invoke, DataType::Type type, CodeGeneratorARM64* codegen) {
     old_value = temps.AcquireSameSizeAs(new_value);
   }
 
-  __ Add(tmp_ptr, base.X(), Operand(offset));
+  ReferenceCodegenARM64<>::DecodeNullable(masm, tmp_ptr, base);
+  __ Add(tmp_ptr, tmp_ptr, Operand(offset));
 
   GenerateCompareAndSet(codegen,
                         type,
@@ -1940,7 +1946,8 @@ static void GenUnsafeGetAndUpdate(HInvoke* invoke,
     codegen->MaybeMarkGCCard(base, /*value=*/arg, new_value_can_be_null);
   }
 
-  __ Add(tmp_ptr, base.X(), Operand(offset));
+  ReferenceCodegenARM64<>::DecodeNullable(masm, tmp_ptr, base);
+  __ Add(tmp_ptr, tmp_ptr, Operand(offset));
   GenerateGetAndUpdate(codegen,
                        get_and_update_op,
                        type,
