diff --git a/compiler/optimizing/intrinsics_arm64.cc b/compiler/optimizing/intrinsics_arm64.cc
--- a/compiler/optimizing/intrinsics_arm64.cc
+++ b/compiler/optimizing/intrinsics_arm64.cc
@@ -4079,13 +4079,14 @@ void IntrinsicCodeGeneratorARM64::VisitCRC32UpdateBytes(HInvoke* invoke) {
   const uint32_t array_data_offset =
       mirror::Array::DataOffset(Primitive::kPrimByte).Uint32Value();
   Register ptr = XRegisterFrom(locations->GetTemp(0));
-  Register array = XRegisterFrom(locations->InAt(1));
+  Register array = WRegisterFrom(locations->InAt(1));
+  ReferenceCodegenARM64<>::DecodeNonNull(masm, ptr, array);
   Location offset = locations->InAt(2);
   if (offset.IsConstant()) {
     int32_t offset_value = offset.GetConstant()->AsIntConstant()->GetValue();
-    __ Add(ptr, array, array_data_offset + offset_value);
+    __ Add(ptr, ptr, array_data_offset + offset_value);
   } else {
-    __ Add(ptr, array, array_data_offset);
+    __ Add(ptr, ptr, array_data_offset);
     __ Add(ptr, ptr, XRegisterFrom(offset));
   }

