diff --git a/compiler/optimizing/intrinsics_arm64.cc b/compiler/optimizing/intrinsics_arm64.cc
--- a/compiler/optimizing/intrinsics_arm64.cc
+++ b/compiler/optimizing/intrinsics_arm64.cc
@@ -3796,11 +3796,13 @@ void IntrinsicCodeGeneratorARM64::VisitReferenceRefersTo(HInvoke* invoke) {
   Register other = WRegisterFrom(locations->InAt(1));
   Register out = WRegisterFrom(locations->Out());
   Register tmp = temps.AcquireW();
+  Register address = out.X();
 
   uint32_t referent_offset = mirror::Reference::ReferentOffset().Uint32Value();
   uint32_t monitor_offset = mirror::Object::MonitorOffset().Int32Value();
 
-  MemOperand field = HeapOperand(obj, referent_offset);
+  ReferenceCodegenARM64<>::DecodeNonNull(masm, address, obj);
+  MemOperand field = HeapOperand(address, referent_offset);
   codegen_->LoadAcquire(invoke, DataType::Type::kReference, tmp, field, /*needs_null_check=*/ true);
   codegen_->GetAssembler()->MaybeUnpoisonHeapReference(tmp);
 
@@ -3822,8 +3824,9 @@ void IntrinsicCodeGeneratorARM64::VisitReferenceRefersTo(HInvoke* invoke) {
 
     // Load the lockword and check if it is a forwarding address.
     static_assert(LockWord::kStateShift == 30u);
     static_assert(LockWord::kStateForwardingAddress == 3u);
-    __ Ldr(tmp, HeapOperand(tmp, monitor_offset));
+    ReferenceCodegenARM64<>::DecodeNonNull(masm, address, tmp);
+    __ Ldr(tmp, HeapOperand(address, monitor_offset));
     __ Cmp(tmp, Operand(0xc0000000));
     __ B(&calculate_result, lo);   // ZF clear if taken.
 
