diff --git a/compiler/optimizing/code_generator_arm64.cc b/compiler/optimizing/code_generator_arm64.cc
--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -7382,6 +7382,12 @@ MemOperand InstructionCodeGeneratorARM64::VecNEONAddress(
   LocationSummary* locations = instruction->GetLocations();
   Register base = InputRegisterAt(instruction, 0);
 
+  if (ReferenceCodegenARM64<>::kNeedsHeapBase) {
+    *scratch = temps_scope->AcquireX();
+    ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), *scratch, base);
+    base = *scratch;
+  }
+
   if (instruction->InputAt(1)->IsIntermediateAddressIndex()) {
     DCHECK(!is_string_char_at);
     return MemOperand(base.X(), InputRegisterAt(instruction, 1).X());
@@ -7399,7 +7405,9 @@ MemOperand InstructionCodeGeneratorARM64::VecNEONAddress(
     offset += Int64FromLocation(index) << shift;
     return HeapOperand(base, offset);
   } else {
-    *scratch = temps_scope->AcquireSameSizeAs(base);
+    if (!scratch->IsValid()) {
+      *scratch = temps_scope->AcquireSameSizeAs(base);
+    }
     __ Add(*scratch, base, Operand(WRegisterFrom(index), LSL, shift));
     return HeapOperand(*scratch, offset);
   }
@@ -7417,5 +7425,11 @@ SVEMemOperand InstructionCodeGeneratorARM64::VecSVEAddress(
   Register base = InputRegisterAt(instruction, 0);
   Location index = locations->InAt(1);
 
+  if (ReferenceCodegenARM64<>::kNeedsHeapBase) {
+    *scratch = temps_scope->AcquireX();
+    ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), *scratch, base);
+    base = *scratch;
+  }
+
   DCHECK(!instruction->InputAt(1)->IsIntermediateAddressIndex());
   DCHECK(!index.IsConstant());
@@ -7430,6 +7444,8 @@ SVEMemOperand InstructionCodeGeneratorARM64::VecSVEAddress(
     return SVEMemOperand(base.X(), XRegisterFrom(index), LSL, shift);
   }
 
-  *scratch = temps_scope->AcquireSameSizeAs(base);
+  if (!scratch->IsValid()) {
+    *scratch = temps_scope->AcquireSameSizeAs(base);
+  }
   __ Add(*scratch, base, offset);
   return SVEMemOperand(scratch->X(), XRegisterFrom(index), LSL, shift);
diff --git a/compiler/optimizing/code_generator_vector_arm64_neon.cc b/compiler/optimizing/code_generator_vector_arm64_neon.cc
--- a/compiler/optimizing/code_generator_vector_arm64_neon.cc
+++ b/compiler/optimizing/code_generator_vector_arm64_neon.cc
@@ -16,6 +16,7 @@
 
 #include "code_generator_arm64.h"
 
+#include "art_reference_codegen_arm64.h"
 #include "arch/arm64/instruction_set_features_arm64.h"
 #include "base/bit_utils_iterator.h"
 #include "mirror/array-inl.h"
@@ -1442,7 +1443,11 @@ void InstructionCodeGeneratorARM64Neon::VisitVecLoad(HVecLoad* instruction) {
                       "Expecting 0=compressed, 1=uncompressed");
         uint32_t count_offset = mirror::String::CountOffset().Uint32Value();
         Register length = temps.AcquireW();
-        __ Ldr(length, HeapOperand(InputRegisterAt(instruction, 0), count_offset));
+        Register native_string = temps.AcquireX();
+        ReferenceCodegenARM64<>::DecodeNonNull(
+            GetVIXLAssembler(), native_string, InputRegisterAt(instruction, 0));
+        __ Ldr(length, MemOperand(native_string, count_offset));
+        temps.Release(native_string);
         __ Tbnz(length.W(), 0, &uncompressed_load);
         temps.Release(length);  // no longer needed
