diff --git a/runtime/gc/heap.cc b/runtime/gc/heap.cc
--- a/runtime/gc/heap.cc
+++ b/runtime/gc/heap.cc
@@ -105,7 +105,9 @@
 #include "verify_object-inl.h"
 #include "well_known_classes.h"
 
-#if defined(__BIONIC__) || defined(__GLIBC__) || defined(ANDROID_HOST_MUSL)
+#if defined(__APPLE__)
+#include <malloc/malloc.h>
+#elif defined(__BIONIC__) || defined(__GLIBC__) || defined(ANDROID_HOST_MUSL)
 #include <malloc.h>  // For mallinfo()
 #endif
 
@@ -2735,7 +2737,11 @@
 
 size_t Heap::GetNativeBytes() {
   size_t malloc_bytes;
-#if defined(__BIONIC__) || defined(__GLIBC__) || defined(ANDROID_HOST_MUSL)
+#if defined(__APPLE__)
+  malloc_statistics_t statistics{};
+  malloc_zone_statistics(malloc_default_zone(), &statistics);
+  malloc_bytes = statistics.size_in_use;
+#elif defined(__BIONIC__) || defined(__GLIBC__) || defined(ANDROID_HOST_MUSL)
   IF_GLIBC(size_t mmapped_bytes;)
   struct mallinfo mi = mallinfo();
   // In spite of the documentation, the jemalloc version of this call seems to do what we want,
