diff --git a/runtime/trace.cc b/runtime/trace.cc
--- a/runtime/trace.cc
+++ b/runtime/trace.cc
@@ -117,6 +117,10 @@ TraceClockSource GetClockSourceFromFlags(int flags) {
   } else if (need_thread_cpu) {
     return TraceClockSource::kThreadCpu;
   } else {
+#if defined(__APPLE__)
+    // This is an Android app runtime, not an ART host tool.
+    return TraceClockSource::kDual;
+#endif
     return kDefaultTraceClockSource;
   }
 }
