diff --git a/runtime/gc/space/image_space.cc b/runtime/gc/space/image_space.cc
--- a/runtime/gc/space/image_space.cc
+++ b/runtime/gc/space/image_space.cc
@@ -207,10 +207,11 @@ class RelocationRange {
  public:
   RelocationRange(const RelocationRange&) = default;
   RelocationRange(uintptr_t source, uintptr_t dest, uintptr_t length)
-      : source_(source),
-        dest_(dest),
+      : source_(ArtHostAddressToCompressedReferenceAddress(source)),
+        dest_(ArtCompressedReferenceToHostAddress(dest)),
         length_(length) {}
 
   bool InSource(uintptr_t address) const {
+    address = ArtHostAddressToCompressedReferenceAddress(address);
     return address - source_ < length_;
   }
@@ -225,6 +228,7 @@ class RelocationRange {
 
   // Translate a source address to the destination space.
   uintptr_t ToDest(uintptr_t address) const {
+    address = ArtHostAddressToCompressedReferenceAddress(address);
     DCHECK(InSource(address));
     return address + Delta();
   }
@@ -2409,7 +2413,7 @@ class ImageSpace::BootImageLoader {
  private:
   class SimpleRelocateVisitor {
    public:
-    SimpleRelocateVisitor(uint32_t diff, uint32_t begin, uint32_t size)
+    SimpleRelocateVisitor(int64_t diff, uint32_t begin, uint32_t size)
         : diff_(diff), begin_(begin), size_(size) {}
 
     // Adapter taking the same arguments as SplitRangeRelocateVisitor
@@ -2417,6 +2417,6 @@ class ImageSpace::BootImageLoader {
     // to simplify constructing the various visitors in DoRelocateSpaces().
-    SimpleRelocateVisitor(uint32_t base_diff,
-                          uint32_t current_diff,
+    SimpleRelocateVisitor(int64_t base_diff,
+                          int64_t current_diff,
                           uint32_t bound,
                           uint32_t begin,
                           uint32_t size)
@@ -2431,25 +2431,27 @@ class ImageSpace::BootImageLoader {
     template <typename T>
     ALWAYS_INLINE T* operator()(T* src) const {
       DCHECK(InSource(src));
-      uint32_t raw_src = reinterpret_cast32<uint32_t>(src);
-      return reinterpret_cast32<T*>(raw_src + diff_);
+      uintptr_t raw_src = ArtHostAddressToCompressedReferenceAddress(
+          reinterpret_cast<uintptr_t>(src));
+      return reinterpret_cast<T*>(static_cast<int64_t>(raw_src) + diff_);
     }
 
     template <typename T>
     ALWAYS_INLINE bool InSource(T* ptr) const {
-      uint32_t raw_ptr = reinterpret_cast32<uint32_t>(ptr);
+      uintptr_t raw_ptr = ArtHostAddressToCompressedReferenceAddress(
+          reinterpret_cast<uintptr_t>(ptr));
       return raw_ptr - begin_ < size_;
     }
 
     template <typename T>
     ALWAYS_INLINE bool InDest(T* ptr) const {
-      uint32_t raw_ptr = reinterpret_cast32<uint32_t>(ptr);
-      uint32_t src_ptr = raw_ptr - diff_;
-      return src_ptr - begin_ < size_;
+      int64_t raw_ptr = static_cast<int64_t>(reinterpret_cast<uintptr_t>(ptr));
+      int64_t src_ptr = raw_ptr - diff_;
+      return src_ptr >= begin_ && static_cast<uint64_t>(src_ptr - begin_) < size_;
     }
 
    private:
-    const uint32_t diff_;
+    const int64_t diff_;
     const uint32_t begin_;
     const uint32_t size_;
   };
@@ -2453,7 +2457,7 @@ class ImageSpace::BootImageLoader {
   class SplitRangeRelocateVisitor {
    public:
-    SplitRangeRelocateVisitor(uint32_t base_diff,
-                              uint32_t current_diff,
+    SplitRangeRelocateVisitor(int64_t base_diff,
+                              int64_t current_diff,
                               uint32_t bound,
                               uint32_t begin,
                               uint32_t size)
@@ -2473,20 +2477,22 @@ class ImageSpace::BootImageLoader {
     template <typename T>
     ALWAYS_INLINE T* operator()(T* src) const {
       DCHECK(InSource(src));
-      uint32_t raw_src = reinterpret_cast32<uint32_t>(src);
-      uint32_t diff = (raw_src < bound_) ? base_diff_ : current_diff_;
-      return reinterpret_cast32<T*>(raw_src + diff);
+      uintptr_t raw_src = ArtHostAddressToCompressedReferenceAddress(
+          reinterpret_cast<uintptr_t>(src));
+      int64_t diff = (raw_src < bound_) ? base_diff_ : current_diff_;
+      return reinterpret_cast<T*>(static_cast<int64_t>(raw_src) + diff);
     }
 
     template <typename T>
     ALWAYS_INLINE bool InSource(T* ptr) const {
-      uint32_t raw_ptr = reinterpret_cast32<uint32_t>(ptr);
+      uintptr_t raw_ptr = ArtHostAddressToCompressedReferenceAddress(
+          reinterpret_cast<uintptr_t>(ptr));
       return raw_ptr - begin_ < size_;
     }
 
    private:
-    const uint32_t base_diff_;
-    const uint32_t current_diff_;
+    const int64_t base_diff_;
+    const int64_t current_diff_;
     const uint32_t bound_;
     const uint32_t begin_;
     const uint32_t size_;
@@ -2540,7 +2545,7 @@ class ImageSpace::BootImageLoader {
     if (base_diff64 == 0 && current_diff64 == 0) {
       return;
     }
-    uint32_t base_diff = static_cast<uint32_t>(base_diff64);
-    uint32_t current_diff = static_cast<uint32_t>(current_diff64);
+    int64_t base_diff = base_diff64;
+    int64_t current_diff = current_diff64;
 
     // For boot image the main visitor is a SimpleRelocateVisitor. For the boot image extension we
@@ -2738,7 +2743,7 @@ class ImageSpace::BootImageLoader {
     ImageSpace* first_space = spaces.front().get();
     const ImageHeader& first_space_header = first_space->GetImageHeader();
     int64_t base_diff64 =
-        static_cast<int64_t>(reinterpret_cast32<uint32_t>(first_space->Begin())) -
+        static_cast<int64_t>(reinterpret_cast<uintptr_t>(first_space->Begin())) -
         static_cast<int64_t>(reinterpret_cast32<uint32_t>(first_space_header.GetImageBegin()));
     if (!relocate_) {
       DCHECK_EQ(base_diff64, 0);
diff --git a/runtime/oat/image.h b/runtime/oat/image.h
--- a/runtime/oat/image.h
+++ b/runtime/oat/image.h
@@ -158,7 +158,8 @@ class PACKED(4) ImageHeader {
   }
 
   uint8_t* GetImageBegin() const {
-    return reinterpret_cast<uint8_t*>(image_begin_);
+    return reinterpret_cast<uint8_t*>(
+        ArtCompressedReferenceToHostAddress(image_begin_));
   }
 
   size_t GetImageSize() const {
@@ -178,19 +179,23 @@ class PACKED(4) ImageHeader {
   // The location that the oat file was expected to be when the image was created. The actual
   // oat file may be at a different location for application images.
   uint8_t* GetOatFileBegin() const {
-    return reinterpret_cast<uint8_t*>(oat_file_begin_);
+    return reinterpret_cast<uint8_t*>(
+        ArtCompressedReferenceToHostAddress(oat_file_begin_));
   }
 
   uint8_t* GetOatDataBegin() const {
-    return reinterpret_cast<uint8_t*>(oat_data_begin_);
+    return reinterpret_cast<uint8_t*>(
+        ArtCompressedReferenceToHostAddress(oat_data_begin_));
   }
 
   uint8_t* GetOatDataEnd() const {
-    return reinterpret_cast<uint8_t*>(oat_data_end_);
+    return reinterpret_cast<uint8_t*>(
+        ArtCompressedReferenceToHostAddress(oat_data_end_));
   }
 
   uint8_t* GetOatFileEnd() const {
-    return reinterpret_cast<uint8_t*>(oat_file_end_);
+    return reinterpret_cast<uint8_t*>(
+        ArtCompressedReferenceToHostAddress(oat_file_end_));
   }
 
   EXPORT PointerSize GetPointerSize() const;

diff --git a/runtime/oat/image-inl.h b/runtime/oat/image-inl.h
--- a/runtime/oat/image-inl.h
+++ b/runtime/oat/image-inl.h
@@ -42,7 +42,8 @@ inline ObjPtr<mirror::ObjectArray<mirror::Object>> ImageHeader::GetImageRoots() const {
   // rather than image_roots_ because it won't move (asserted below)
   // and it's a const member.
   mirror::ObjectArray<mirror::Object>* image_roots =
-      reinterpret_cast<mirror::ObjectArray<mirror::Object>*>(image_roots_);
+      reinterpret_cast<mirror::ObjectArray<mirror::Object>*>(
+          ArtCompressedReferenceToHostAddress(image_roots_));
   mirror::ObjectArray<mirror::Object>* result =
       ReadBarrier::BarrierForRoot<mirror::ObjectArray<mirror::Object>, kReadBarrierOption>(
           &image_roots);
diff --git a/runtime/gc/heap.cc b/runtime/gc/heap.cc
--- a/runtime/gc/heap.cc
+++ b/runtime/gc/heap.cc
@@ -4530,21 +4530,29 @@ bool Heap::IsGCDisabledForShutdown() const {
 }
 
 bool Heap::ObjectIsInBootImageSpace(ObjPtr<mirror::Object> obj) const {
-  DCHECK_EQ(IsBootImageAddress(obj.Ptr()),
-            any_of(boot_image_spaces_.begin(),
-                   boot_image_spaces_.end(),
-                   [obj](gc::space::ImageSpace* space) REQUIRES_SHARED(Locks::mutator_lock_) {
-                     return space->HasAddress(obj.Ptr());
-                   }));
-  return IsBootImageAddress(obj.Ptr());
+  bool is_boot_image_address =
+      ArtHostAddressToCompressedReferenceAddress(reinterpret_cast<uintptr_t>(obj.Ptr())) -
+          boot_images_start_address_ < boot_images_size_;
+  bool is_in_space = any_of(
+      boot_image_spaces_.begin(),
+      boot_image_spaces_.end(),
+      [obj](gc::space::ImageSpace* space) REQUIRES_SHARED(Locks::mutator_lock_) {
+        return space->HasAddress(obj.Ptr());
+      });
+  DCHECK_EQ(is_boot_image_address, is_in_space);
+  return is_boot_image_address;
 }
 
 bool Heap::IsInBootImageOatFile(const void* p) const {
-  DCHECK_EQ(IsBootImageAddress(p),
-            any_of(boot_image_spaces_.begin(),
-                   boot_image_spaces_.end(),
-                   [p](gc::space::ImageSpace* space) REQUIRES_SHARED(Locks::mutator_lock_) {
-                     return space->GetOatFile()->Contains(p);
-                   }));
-  return IsBootImageAddress(p);
+  bool is_boot_image_address =
+      ArtHostAddressToCompressedReferenceAddress(reinterpret_cast<uintptr_t>(p)) -
+          boot_images_start_address_ < boot_images_size_;
+  bool is_in_oat_file = any_of(
+      boot_image_spaces_.begin(),
+      boot_image_spaces_.end(),
+      [p](gc::space::ImageSpace* space) REQUIRES_SHARED(Locks::mutator_lock_) {
+        return space->GetOatFile()->Contains(p);
+      });
+  DCHECK_EQ(is_boot_image_address, is_in_oat_file);
+  return is_boot_image_address;
 }
@@ -4857,7 +4865,11 @@ bool Heap::HasAppImageSpaceFor(const std::string& dex_location) const {
   for (space::ContinuousSpace* space : continuous_spaces_) {
     // An image space is either a boot image space or an app image space.
     if (space->IsImageSpace() &&
-        !IsBootImageAddress(space->Begin()) &&
+        !any_of(boot_image_spaces_.begin(),
+                boot_image_spaces_.end(),
+                [space](gc::space::ImageSpace* image) {
+                  return image->Begin() == space->Begin();
+                }) &&
         (space->AsImageSpace()->GetOatFile()->GetOatDexFiles()[0]->GetDexFileLocation() ==
               dex_location)) {
       return true;
diff --git a/runtime/class_linker.cc b/runtime/class_linker.cc
--- a/runtime/class_linker.cc
+++ b/runtime/class_linker.cc
@@ -11305,7 +11305,10 @@ ObjPtr<mirror::ClassLoader> ClassLinker::GetHoldingClassLoaderOfCopiedMethod(
   CHECK(method->IsCopied());
   gc::Heap* heap = Runtime::Current()->GetHeap();
   // Check if the copied method is in the boot class path.
-  if (heap->IsBootImageAddress(method) || GetAllocatorForClassLoader(nullptr)->Contains(method)) {
+  uintptr_t logical_method =
+      ArtHostAddressToCompressedReferenceAddress(reinterpret_cast<uintptr_t>(method));
+  if (logical_method - heap->GetBootImagesStartAddress() < heap->GetBootImagesSize() ||
+      GetAllocatorForClassLoader(nullptr)->Contains(method)) {
     DCHECK(GetHoldingClassOfCopiedMethod(method)->GetClassLoader() == nullptr);
     return nullptr;
   }
