--- a/runtime/native_bridge_art_interface.cc
+++ b/runtime/native_bridge_art_interface.cc
@@ -101,14 +101,12 @@ void PreInitializeNativeBridge(const std::string& dir) {
   VLOG(startup) << "Runtime::Pre-initialize native bridge";
 #ifndef __APPLE__  // Mac OS does not support CLONE_NEWNS.
   if (unshare(CLONE_NEWNS) == -1) {
     LOG(WARNING) << "Could not create mount namespace.";
   }
-  android::PreInitializeNativeBridge(dir.c_str(), GetInstructionSetString(kRuntimeISA));
-#else
-  UNUSED(dir);
 #endif
+  android::PreInitializeNativeBridge(dir.c_str(), GetInstructionSetString(kRuntimeISA));
 }

 void PreZygoteForkNativeBridge() {
