diff --git a/runtime/runtime.cc b/runtime/runtime.cc
--- a/runtime/runtime.cc
+++ b/runtime/runtime.cc
@@ -1071,9 +1071,15 @@ void Runtime::FinishMinimalForDarwinProbe() {
   // configure/start the metrics reporter, and publish the Signal Catcher.
   if (!is_zygote_) {
+    if (is_native_bridge_loaded_) {
+      PreInitializeNativeBridge(".");
+    }
+    NativeBridgeAction action = force_native_bridge_
+        ? NativeBridgeAction::kInitialize
+        : NativeBridgeAction::kUnload;
     InitNonZygoteOrPostFork(self->GetJniEnv(),
                             /*is_system_server=*/false,
                             /*is_child_zygote=*/false,
-                            NativeBridgeAction::kUnload,
+                            action,
                             GetInstructionSetString(kRuntimeISA));
   }
   {
