diff --git a/compiler/optimizing/code_generator_arm64.cc b/compiler/optimizing/code_generator_arm64.cc
--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -4997,6 +4997,8 @@
       // Note: Boot image is in the low 4GiB and the entry is 32-bit, so emit a 32-bit load.
       uint32_t boot_image_offset = GetBootImageOffset(invoke);
       LoadBootImageRelRoEntry(WRegisterFrom(temp), boot_image_offset);
+      ReferenceCodegenARM64<>::DecodeNonNull(
+          GetVIXLAssembler(), XRegisterFrom(temp), WRegisterFrom(temp));
       break;
     }
     case MethodLoadKind::kAppImageRelRo: {
@@ -5010,6 +5012,8 @@
       vixl::aarch64::Label* ldr_label =
           NewAppImageMethodPatch(invoke->GetResolvedMethodReference(), adrp_label);
       EmitLdrOffsetPlaceholder(ldr_label, WRegisterFrom(temp), XRegisterFrom(temp));
+      ReferenceCodegenARM64<>::DecodeNonNull(
+          GetVIXLAssembler(), XRegisterFrom(temp), WRegisterFrom(temp));
       break;
     }
     case MethodLoadKind::kBssEntry: {
