diff --git a/compiler/optimizing/code_generator_arm64.cc b/compiler/optimizing/code_generator_arm64.cc
--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -6439,10 +6439,21 @@ void CodeGeneratorARM64::GenerateImplicitNullCheck(HNullCheck* instruction) {
     return;
   }
   {
+    MacroAssembler* masm = GetVIXLAssembler();
+    UseScratchRegisterScope temps(masm);
     // Ensure that between load and RecordPcInfo there are no pools emitted.
-    EmissionCheckScope guard(GetVIXLAssembler(), kMaxMacroInstructionSizeInBytes);
+    EmissionCheckScope guard(masm, 2 * kMaxMacroInstructionSizeInBytes);
     Location obj = instruction->GetLocations()->InAt(0);
-    __ Ldr(wzr, HeapOperandFrom(obj, Offset(0)));
+    if constexpr (ReferenceCodegenARM64<>::kNeedsHeapBase) {
+      Register address = temps.AcquireX();
+      ReferenceCodegenARM64<>::DecodeNullable(
+          masm,
+          address,
+          RegisterFrom(obj, DataType::Type::kReference).W());
+      __ Ldr(wzr, MemOperand(address));
+    } else {
+      __ Ldr(wzr, HeapOperandFrom(obj, Offset(0)));
+    }
     RecordPcInfo(instruction);
   }
 }
