diff --git a/compiler/optimizing/intrinsics_arm64.cc b/compiler/optimizing/intrinsics_arm64.cc
--- a/compiler/optimizing/intrinsics_arm64.cc
+++ b/compiler/optimizing/intrinsics_arm64.cc
@@ -3713,6 +3713,7 @@ void IntrinsicCodeGeneratorARM64::HandleValueOf(HInvoke* invoke,
   auto allocate_instance = [&]() {
     DCHECK(out.X().Is(InvokeRuntimeCallingConvention().GetRegisterAt(0)));
     codegen_->LoadIntrinsicDeclaringClass(out, invoke);
+    ReferenceCodegenARM64<>::DecodeNonNull(masm, out.X(), out.W());
     codegen_->InvokeRuntime(kQuickAllocObjectInitialized, invoke);
     CheckEntrypointTypes<kQuickAllocObjectWithChecks, void*, mirror::Class*>();
   };
@@ -3727,10 +3728,11 @@ void IntrinsicCodeGeneratorARM64::HandleValueOf(HInvoke* invoke,
       // TODO: If we JIT, we could allocate the object now, and store it in the
       // JIT object table.
       allocate_instance();
       __ Mov(temp.W(), value);
-      codegen_->Store(type, temp.W(), HeapOperand(out.W(), info.value_offset));
+      codegen_->Store(type, temp.W(), HeapOperand(out.X(), info.value_offset));
       // Class pointer and `value` final field stores require a barrier before publication.
       codegen_->GenerateMemoryBarrier(MemBarrierKind::kStoreStore);
+      ReferenceCodegenARM64<>::Encode(masm, out.X(), out.X());
     }
   } else {
     DCHECK(locations->CanCall());
@@ -3750,9 +3752,10 @@ void IntrinsicCodeGeneratorARM64::HandleValueOf(HInvoke* invoke,
     __ Bind(&allocate);
     // Otherwise allocate and initialize a new object.
     allocate_instance();
-    codegen_->Store(type, in.W(), HeapOperand(out.W(), info.value_offset));
+    codegen_->Store(type, in.W(), HeapOperand(out.X(), info.value_offset));
     // Class pointer and `value` final field stores require a barrier before publication.
     codegen_->GenerateMemoryBarrier(MemBarrierKind::kStoreStore);
+    ReferenceCodegenARM64<>::Encode(masm, out.X(), out.X());
     __ Bind(&done);
   }
 }
