diff --git a/compiler/optimizing/code_generator_arm64.cc b/compiler/optimizing/code_generator_arm64.cc
--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -5019,7 +5019,11 @@ void InstructionCodeGeneratorARM64::VisitInvokeInterface(HInvokeInterface* invoke) {
   // Ensure that between load and MaybeRecordImplicitNullCheck there are no pools emitted.
   if (receiver.IsStackSlot()) {
     __ Ldr(temp.W(), StackOperandFrom(receiver));
-    ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), temp, temp);
+    if (codegen_->GetCompilerOptions().GetImplicitNullChecks()) {
+      ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), temp, temp.W());
+    } else {
+      ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), temp, temp.W());
+    }
     {
       EmissionCheckScope guard(GetVIXLAssembler(), kMaxMacroInstructionSizeInBytes);
       // /* HeapReference<Class> */ temp = temp->klass_
@@ -5027,7 +5031,13 @@ void InstructionCodeGeneratorARM64::VisitInvokeInterface(HInvokeInterface* invoke) {
       codegen_->MaybeRecordImplicitNullCheck(invoke);
     }
   } else {
-    ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), temp, WRegisterFrom(receiver));
+    if (codegen_->GetCompilerOptions().GetImplicitNullChecks()) {
+      ReferenceCodegenARM64<>::DecodeNullable(
+          GetVIXLAssembler(), temp, WRegisterFrom(receiver));
+    } else {
+      ReferenceCodegenARM64<>::DecodeNonNull(
+          GetVIXLAssembler(), temp, WRegisterFrom(receiver));
+    }
     EmissionCheckScope guard(GetVIXLAssembler(), kMaxMacroInstructionSizeInBytes);
     // /* HeapReference<Class> */ temp = receiver->klass_
     __ Ldr(temp.W(), MemOperand(temp, class_offset.SizeValue()));
@@ -5345,8 +5355,11 @@ void CodeGeneratorARM64::GenerateVirtualCall(
 
   MemOperand class_address = HeapOperandFrom(LocationFrom(receiver), class_offset);
   if (ReferenceCodegenARM64<>::kNeedsHeapBase) {
-    DCHECK(!GetCompilerOptions().GetImplicitNullChecks());
-    ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), temp, receiver);
+    if (GetCompilerOptions().GetImplicitNullChecks()) {
+      ReferenceCodegenARM64<>::DecodeNullable(GetVIXLAssembler(), temp, receiver.W());
+    } else {
+      ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), temp, receiver.W());
+    }
     class_address = MemOperand(temp, class_offset.SizeValue());
   }
   {
