diff --git a/compiler/optimizing/code_generator_arm64.cc b/compiler/optimizing/code_generator_arm64.cc
--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -2369,9 +2369,13 @@ void InstructionCodeGeneratorARM64::HandleFieldGet(HInstruction* instruction,
   MemOperand field = HeapOperand(InputRegisterAt(instruction, receiver_input), field_info.GetFieldOffset());
   if (ReferenceCodegenARM64<>::kNeedsHeapBase) {
-    DCHECK(!codegen_->GetCompilerOptions().GetImplicitNullChecks());
     native_holder = XRegisterFrom(locations->GetTemp(0));
-    ReferenceCodegenARM64<>::DecodeNonNull(
-        GetVIXLAssembler(), native_holder, holder);
+    if (codegen_->GetCompilerOptions().GetImplicitNullChecks()) {
+      ReferenceCodegenARM64<>::DecodeNullable(
+          GetVIXLAssembler(), native_holder, holder.W());
+    } else {
+      ReferenceCodegenARM64<>::DecodeNonNull(
+          GetVIXLAssembler(), native_holder, holder.W());
+    }
     field = MemOperand(native_holder, offset);
   }
 
@@ -2491,10 +2495,14 @@ void InstructionCodeGeneratorARM64::HandleFieldSet(HInstruction* instruction,
   if (ReferenceCodegenARM64<>::kNeedsHeapBase) {
     LocationSummary* locations = instruction->GetLocations();
     Register address = XRegisterFrom(locations->GetTemp(0));
-    ReferenceCodegenARM64<>::DecodeNonNull(
-        GetVIXLAssembler(),
-        address,
-        RegisterFrom(locations->InAt(0), DataType::Type::kReference));
+    Register receiver = RegisterFrom(locations->InAt(0), DataType::Type::kReference);
+    if (codegen_->GetCompilerOptions().GetImplicitNullChecks()) {
+      ReferenceCodegenARM64<>::DecodeNullable(
+          GetVIXLAssembler(), address, receiver.W());
+    } else {
+      ReferenceCodegenARM64<>::DecodeNonNull(
+          GetVIXLAssembler(), address, receiver.W());
+    }
     destination = MemOperand(address, offset.Uint32Value());
   }
   DataType::Type field_type = field_info.GetFieldType();
