--- a/compiler/optimizing/intrinsics_arm64.cc
+++ b/compiler/optimizing/intrinsics_arm64.cc
@@ -3094,6 +3094,7 @@ void IntrinsicLocationsBuilderARM64::VisitSystemArrayCopyChar(HInvoke* invoke) {
   locations->SetInAt(4, LocationForSystemArrayCopyInput(invoke->InputAt(4)));
 
   locations->AddRegisterTemps(3);
+  locations->AddRegisterTemps(2);
 }
 
 static void CheckSystemArrayCopyPosition(MacroAssembler* masm,
@@ -3100,10 +3101,12 @@ void IntrinsicCodeGeneratorARM64::VisitSystemArrayCopyChar(HInvoke* invoke) {
   MacroAssembler* masm = GetVIXLAssembler();
   LocationSummary* locations = invoke->GetLocations();
-  Register src = XRegisterFrom(locations->InAt(0));
+  Register src_ref = WRegisterFrom(locations->InAt(0));
   Location src_pos = locations->InAt(1);
-  Register dst = XRegisterFrom(locations->InAt(2));
+  Register dst_ref = WRegisterFrom(locations->InAt(2));
   Location dst_pos = locations->InAt(3);
   Location length = locations->InAt(4);
+  Register src = XRegisterFrom(locations->GetTemp(3));
+  Register dst = XRegisterFrom(locations->GetTemp(4));
 
   SlowPathCodeARM64* slow_path =
       new (codegen_->GetScopedAllocator()) IntrinsicSlowPathARM64(invoke);
@@ -3112,14 +3115,17 @@ void IntrinsicCodeGeneratorARM64::VisitSystemArrayCopyChar(HInvoke* invoke) {
 
   // If source and destination are the same, take the slow path. Overlapping copy regions must be
   // copied in reverse and we can't know in all cases if it's needed.
-  __ Cmp(src, dst);
+  __ Cmp(src_ref, dst_ref);
   __ B(slow_path->GetEntryLabel(), eq);
 
   // Bail out if the source is null.
-  __ Cbz(src, slow_path->GetEntryLabel());
+  __ Cbz(src_ref, slow_path->GetEntryLabel());
 
   // Bail out if the destination is null.
-  __ Cbz(dst, slow_path->GetEntryLabel());
+  __ Cbz(dst_ref, slow_path->GetEntryLabel());
+
+  ReferenceCodegenARM64<>::DecodeNonNull(masm, src, src_ref);
+  ReferenceCodegenARM64<>::DecodeNonNull(masm, dst, dst_ref);
 
   if (!length.IsConstant()) {
     // Merge the following two comparisons into one:
