diff --git a/compiler/optimizing/intrinsics_arm64.cc b/compiler/optimizing/intrinsics_arm64.cc
--- a/compiler/optimizing/intrinsics_arm64.cc
+++ b/compiler/optimizing/intrinsics_arm64.cc
@@ -3760,7 +3760,10 @@ void IntrinsicCodeGeneratorARM64::VisitReferenceGetReferent(HInvoke* invoke) {
     DCHECK_ALIGNED(disable_intrinsic_offset.Uint32Value(), 2u);
     DCHECK_EQ(disable_intrinsic_offset.Uint32Value() + 1u,
               IntrinsicVisitor::GetReferenceSlowPathEnabledOffset().Uint32Value());
-    __ Ldrh(temp, HeapOperand(temp, disable_intrinsic_offset.Uint32Value()));
+    ReferenceCodegenARM64<>::DecodeNonNull(
+        GetVIXLAssembler(), temp.X(), temp.W());
+    __ Ldrh(temp,
+            MemOperand(temp.X(), disable_intrinsic_offset.Uint32Value()));
     __ Cbnz(temp, slow_path->GetEntryLabel());
   }
 
