diff --git a/compiler/jni/quick/jni_compiler.cc b/compiler/jni/quick/jni_compiler.cc
--- a/compiler/jni/quick/jni_compiler.cc
+++ b/compiler/jni/quick/jni_compiler.cc
@@ -269,11 +269,15 @@ static JniCompiledMethod ArtJniCompileMethodInternal(const CompilerOptions& compiler_options,
   std::unique_ptr<JNIMacroLabel> transition_to_native_slow_path;
   std::unique_ptr<JNIMacroLabel> transition_to_native_resume;
   if (LIKELY(!is_critical_native && !is_fast_native)) {
+#if defined(__APPLE__)
+    __ CallFromThread(QUICK_ENTRYPOINT_OFFSET(kPointerSize, pJniMethodStart));
+#else
     transition_to_native_slow_path = __ CreateLabel();
     transition_to_native_resume = __ CreateLabel();
     __ TryToTransitionFromRunnableToNative(transition_to_native_slow_path.get(),
                                            callee_save_scratch_regs);
     __ Bind(transition_to_native_resume.get());
+#endif
   }
 
   // 3. Push local reference frame.
@@ -470,12 +474,16 @@ static JniCompiledMethod ArtJniCompileMethodInternal(const CompilerOptions& compiler_options,
   std::unique_ptr<JNIMacroLabel> transition_to_runnable_slow_path;
   std::unique_ptr<JNIMacroLabel> transition_to_runnable_resume;
   if (LIKELY(!is_critical_native && !is_fast_native)) {
+#if defined(__APPLE__)
+    __ CallFromThread(QUICK_ENTRYPOINT_OFFSET(kPointerSize, pJniMethodEnd));
+#else
     transition_to_runnable_slow_path = __ CreateLabel();
     transition_to_runnable_resume = __ CreateLabel();
     __ TryToTransitionFromNativeToRunnable(transition_to_runnable_slow_path.get(),
                                            main_jni_conv->ArgumentScratchRegisters(),
                                            mr_conv->ReturnRegister());
     __ Bind(transition_to_runnable_resume.get());
+#endif
   }
 
   // 5.2. For methods that return a reference, do an exception check before decoding the reference.
@@ -632,15 +640,19 @@ static JniCompiledMethod ArtJniCompileMethodInternal(const CompilerOptions& compiler_options,
   // 8.2. Slow path for transition to Native.
   if (LIKELY(!is_critical_native && !is_fast_native)) {
+#if !defined(__APPLE__)
     __ Bind(transition_to_native_slow_path.get());
     __ CallFromThread(QUICK_ENTRYPOINT_OFFSET(kPointerSize, pJniMethodStart));
     __ Jump(transition_to_native_resume.get());
+#endif
   }
 
   // 8.3. Slow path for transition to Runnable.
   if (LIKELY(!is_critical_native && !is_fast_native)) {
+#if !defined(__APPLE__)
     __ Bind(transition_to_runnable_slow_path.get());
     __ CallFromThread(QUICK_ENTRYPOINT_OFFSET(kPointerSize, pJniMethodEnd));
     __ Jump(transition_to_runnable_resume.get());
+#endif
   }
 
   // 8.4. Exception poll slow path(s).
