diff --git a/runtime/gc/space/image_space.cc b/runtime/gc/space/image_space.cc
--- a/runtime/gc/space/image_space.cc
+++ b/runtime/gc/space/image_space.cc
@@ -2886,3 +2886,5 @@
       std::string oat_filename =
           ImageHeader::GetOatLocationFromImageLocation(space->GetImageFilename());
+      std::string oat_location =
+          ImageHeader::GetOatLocationFromImageLocation(space->GetImageLocation());
 
@@ -2895,4 +2897,4 @@
                                      oat_filename,
-                                     oat_filename,
+                                     oat_location,
                                      executable_,
                                      /*low_4gb=*/false,
@@ -2903,4 +2905,4 @@
                                      vdex_fd.get(),
                                      oat_fd.get(),
-                                     oat_filename,
+                                     oat_location,
                                      executable_,
                                      /*low_4gb=*/false,
