diff --git a/runtime/obj_ptr-inl.h b/runtime/obj_ptr-inl.h
--- a/runtime/obj_ptr-inl.h
+++ b/runtime/obj_ptr-inl.h
@@ -54,6 +54,13 @@
 inline uintptr_t ObjPtr<MirrorType>::Encode(MirrorType* ptr) {
   uintptr_t ref = reinterpret_cast<uintptr_t>(ptr);
   DCHECK_ALIGNED(ref, kObjectAlignment);
+#if defined(__APPLE__)
+  if (!kObjPtrPoisoning && ref != 0u) {
+    DCHECK_GE(ref, kArtCompressedReferenceBase);
+    DCHECK_LT(ref, kArtCompressedReferenceLimit);
+    return ref - kArtCompressedReferenceBase;
+  }
+#endif
   if (kObjPtrPoisoning && ref != 0) {
     DCHECK_LE(ref, 0xFFFFFFFFU);
     ref >>= kObjectAlignmentShift;
diff --git a/runtime/obj_ptr.h b/runtime/obj_ptr.h
--- a/runtime/obj_ptr.h
+++ b/runtime/obj_ptr.h
@@ -94,4 +94,12 @@
       return reinterpret_cast<MirrorType*>(
           static_cast<uintptr_t>(static_cast<uint32_t>(reference_ << kObjectAlignmentShift)));
     } else {
+#if defined(__APPLE__)
+      if (reference_ != 0u) {
+        if (reference_ >= kArtCompressedReferenceBase) {
+          return reinterpret_cast<MirrorType*>(reference_);
+        }
+        return reinterpret_cast<MirrorType*>(kArtCompressedReferenceBase + reference_);
+      }
+#endif
       return reinterpret_cast<MirrorType*>(reference_);
     }
   }
