diff --git a/runtime/art_method.h b/runtime/art_method.h
--- a/runtime/art_method.h
+++ b/runtime/art_method.h
@@ -790,5 +790,12 @@
   ALWAYS_INLINE
   const void* GetEntryPointFromQuickCompiledCodePtrSize(PointerSize pointer_size) const {
-    return GetNativePointer<const void*>(
-        EntryPointFromQuickCompiledCodeOffset(pointer_size), pointer_size);
+    const void* entry_point = GetNativePointer<const void*>(
+        EntryPointFromQuickCompiledCodeOffset(pointer_size), pointer_size);
+#if defined(__APPLE__)
+    if (pointer_size == PointerSize::k32) {
+      entry_point = reinterpret_cast<const void*>(ArtCompressedReferenceToHostAddress(
+          reinterpret_cast<uintptr_t>(entry_point)));
+    }
+#endif
+    return entry_point;
  }
