diff --git a/runtime/entrypoints/quick/quick_trampoline_entrypoints.cc b/runtime/entrypoints/quick/quick_trampoline_entrypoints.cc
--- a/runtime/entrypoints/quick/quick_trampoline_entrypoints.cc
+++ b/runtime/entrypoints/quick/quick_trampoline_entrypoints.cc
@@ -2071,4 +2071,13 @@
   // Fix up managed-stack things in Thread. After this we can walk the stack.
   self->SetTopOfStackGenericJniTagged(managed_sp);
 
+#if defined(__APPLE__)
+  uint32_t caller_dex_pc = 0;
+  ArtMethod* caller_method = QuickArgumentVisitor::GetCallingMethodAndDexPc(
+      managed_sp, &caller_dex_pc);
+  if (caller_method != nullptr) {
+    caller_method->GetEntryPointFromQuickCompiledCode();
+  }
+#endif
+
   self->VerifyStack();
