diff --git a/runtime/art_method.cc b/runtime/art_method.cc
--- a/runtime/art_method.cc
+++ b/runtime/art_method.cc
@@ -630,8 +630,16 @@
   jit::Jit* jit = runtime->GetJit();
   if (jit != nullptr) {
     jit::JitCodeCache* code_cache = jit->GetCodeCache();
     OatQuickMethodHeader* method_header = code_cache->LookupMethodHeader(pc, this);
     if (method_header != nullptr) {
       DCHECK(method_header->Contains(pc));
+#if defined(__APPLE__)
+      darwin_art_register_compiled_method(
+          reinterpret_cast<uintptr_t>(method_header->GetCode()), method_header->GetCodeSize(),
+          PrettyMethod(/*with_signature=*/false).c_str());
+      DarwinArtRegisterJitMethod(reinterpret_cast<uintptr_t>(method_header->GetCode()),
+                                 method_header->GetCodeSize(),
+                                 reinterpret_cast<uintptr_t>(this));
+#endif
       return method_header;

diff --git a/runtime/art_method.h b/runtime/art_method.h
--- a/runtime/art_method.h
+++ b/runtime/art_method.h
@@ -24,3 +24,4 @@
 extern "C" void darwin_art_register_compiled_method(uintptr_t start, size_t size,
                                                       const char* name);
+extern "C" void DarwinArtRegisterJitMethod(uintptr_t code, size_t size, uintptr_t method);
 #endif
