diff --git a/runtime/art_method.cc b/runtime/art_method.cc
--- a/runtime/art_method.cc
+++ b/runtime/art_method.cc
@@ -695,2 +695,10 @@
   OatQuickMethodHeader* method_header = OatQuickMethodHeader::FromEntryPoint(oat_entry_point);
+#if defined(__APPLE__)
+  darwin_art_register_compiled_method(
+      reinterpret_cast<uintptr_t>(method_header->GetCode()), method_header->GetCodeSize(),
+      PrettyMethod(/*with_signature=*/false).c_str());
+  DarwinArtRegisterJitMethod(reinterpret_cast<uintptr_t>(method_header->GetCode()),
+                             method_header->GetCodeSize(),
+                             reinterpret_cast<uintptr_t>(this));
+#endif
   // We could have existing Oat code for native methods but we may not use it if the runtime is java
