diff --git a/runtime/class_linker.cc b/runtime/class_linker.cc
--- a/runtime/class_linker.cc
+++ b/runtime/class_linker.cc
@@ -499,5 +499,10 @@ const void* ClassLinker::RegisterNative(
   } else {
     method->SetEntryPointFromJni(new_native_method);
   }
+#if defined(__APPLE__)
+  darwin_art_register_native_method(
+      reinterpret_cast<uintptr_t>(new_native_method),
+      method->PrettyMethod(/*with_signature=*/false).c_str());
+#endif
  return new_native_method;
 }
