--- a/compiler/optimizing/code_generator_arm64.cc
+++ b/compiler/optimizing/code_generator_arm64.cc
@@ -2216,6 +2216,13 @@
                                                                      Register class_reg) {
   UseScratchRegisterScope temps(GetVIXLAssembler());
   Register temp = temps.AcquireW();
+  if (ReferenceCodegenARM64<>::kNeedsHeapBase) {
+    // Decode compressed class references before reading status.
+    temps.Exclude(class_reg);
+    Register native_class = temps.AcquireX();
+    ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), native_class, class_reg.W());
+    class_reg = native_class;
+  }
 
   // CMP (immediate) is limited to imm12 or imm12<<12, so we would need to materialize
