--- a/compiler/optimizing/fast_compiler_arm64.cc
+++ b/compiler/optimizing/fast_compiler_arm64.cc
@@ -919,7 +919,11 @@
       Register receiver = calling_convention.GetRegisterAt(0);
       Offset class_offset = mirror::Object::ClassOffset();
       EmissionCheckScope guard(GetVIXLAssembler(), kMaxMacroInstructionSizeInBytes);
-      __ Ldr(kArtMethodRegister.W(), HeapOperand(receiver.W(), class_offset));
+      UseScratchRegisterScope temps(GetVIXLAssembler());
+      temps.Exclude(receiver);
+      Register native_receiver = temps.AcquireX();
+      ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), native_receiver, receiver.W());
+      __ Ldr(kArtMethodRegister.W(), MemOperand(native_receiver, class_offset));
       if (can_be_null) {
