--- a/compiler/optimizing/fast_compiler_arm64.cc
+++ b/compiler/optimizing/fast_compiler_arm64.cc
@@ -27,6 +27,7 @@
 
 #include "code_generation_data.h"
 #include "code_generator_arm64.h"
+#include "art_reference_codegen_arm64.h"
 #include "data_type-inl.h"
 #include "dex/code_item_accessors-inl.h"
 #include "dex/dex_instruction-inl.h"
@@ -923,7 +923,7 @@
       temps.Exclude(receiver);
       Register native_receiver = temps.AcquireX();
       ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), native_receiver, receiver.W());
-      __ Ldr(kArtMethodRegister.W(), MemOperand(native_receiver, class_offset));
+      __ Ldr(kArtMethodRegister.W(), MemOperand(native_receiver, class_offset.Int32Value()));
       if (can_be_null) {
         RecordPcInfo(dex_pc);
