diff --git a/runtime/class_linker.cc b/runtime/class_linker.cc
--- a/runtime/class_linker.cc
+++ b/runtime/class_linker.cc
@@ -1471,4 +1471,26 @@ bool ClassLinker::InitFromBootImage(std::string* error_msg) {
   InitializeObjectVirtualMethodHashes(GetClassRoot<mirror::Object>(this),
                                       image_pointer_size_,
                                       ArrayRef<uint32_t>(object_virtual_method_hashes_));
+#if defined(__APPLE__)
+  class DarwinBootOatRegistryVisitor final : public ClassVisitor {
+   public:
+    bool operator()(ObjPtr<mirror::Class> klass) override
+        REQUIRES_SHARED(Locks::mutator_lock_) {
+      for (ArtMethod& method : klass->GetMethods(kRuntimePointerSize)) {
+        if (method.IsAbstract() || method.IsNative()) {
+          continue;
+        }
+        const void* entrypoint = method.GetOatMethodQuickCode(kRuntimePointerSize);
+        if (entrypoint == nullptr) {
+          continue;
+        }
+        const OatQuickMethodHeader* header = OatQuickMethodHeader::FromEntryPoint(entrypoint);
+        DarwinArtRegisterJitMethod(reinterpret_cast<uintptr_t>(header->GetCode()),
+                                   header->GetCodeSize(), reinterpret_cast<uintptr_t>(&method));
+      }
+      return true;
+    }
+  } visitor;
+  VisitClasses(&visitor);
+#endif
   FinishInit(self);
