--- a/compiler/optimizing/fast_compiler_arm64.cc
+++ b/compiler/optimizing/fast_compiler_arm64.cc
@@ -1925,10 +1925,14 @@
         }
       }
 
-      MemOperand mem = HeapOperand(
-          RegisterFrom(GetExistingRegisterLocation(obj_reg, DataType::Type::kReference),
-                       DataType::Type::kReference),
-          field->GetOffset());
+      UseScratchRegisterScope address_temps(GetVIXLAssembler());
+      Register holder = RegisterFrom(
+          GetExistingRegisterLocation(obj_reg, DataType::Type::kReference),
+          DataType::Type::kReference);
+      address_temps.Exclude(holder);
+      Register native_holder = address_temps.AcquireX();
+      ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), native_holder, holder.W());
+      MemOperand mem = MemOperand(native_holder, field->GetOffset().Int32Value());
       if (HitUnimplemented()) {
         return false;
       }
@@ -2047,7 +2051,11 @@
       if (HitUnimplemented()) {
         return false;
       }
-      MemOperand mem = HeapOperand(holder, field->GetOffset());
+      UseScratchRegisterScope address_temps(GetVIXLAssembler());
+      address_temps.Exclude(holder);
+      Register native_holder = address_temps.AcquireX();
+      ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), native_holder, holder.W());
+      MemOperand mem = MemOperand(native_holder, field->GetOffset().Int32Value());
 
       // Need one temp if the stored value is a constant.
       UseScratchRegisterScope temps(GetVIXLAssembler());
