--- a/compiler/optimizing/fast_compiler_arm64.cc
+++ b/compiler/optimizing/fast_compiler_arm64.cc
@@ -1075,6 +1075,10 @@
   Register cls = calling_convention.GetRegisterAt(1);
   Register obj_cls = calling_convention.GetRegisterAt(2);
   Register obj = WRegisterFrom(GetExistingRegisterLocation(vreg, DataType::Type::kReference));
+  temps.Exclude(cls);
+  temps.Exclude(obj_cls);
+  temps.Exclude(obj.X());
+  Register native_obj = temps.AcquireX();
   if (HitUnimplemented()) {
     return false;
   }
@@ -1091,12 +1095,13 @@
 
   vixl::aarch64::Label exit, read_barrier_exit;
   __ Cbz(obj, &exit);
+  ReferenceCodegenARM64<>::DecodeNonNull(GetVIXLAssembler(), native_obj, obj.W());
   __ Ldr(cls.X(), jit_patches_.DeduplicateJitClassLiteral(GetDexFile(),
                                                           type_index,
                                                           h_klass ,
                                                           code_generation_data_.get()));
   __ Ldr(cls.W(), MemOperand(cls.X()));
-  __ Ldr(obj_cls.W(), MemOperand(obj.X()));
+  __ Ldr(obj_cls.W(), MemOperand(native_obj));
   __ Cmp(cls.W(), obj_cls.W());
   __ B(eq, &exit);
 
