diff --git a/compiler/optimizing/intrinsics_arm64.cc b/compiler/optimizing/intrinsics_arm64.cc
--- a/compiler/optimizing/intrinsics_arm64.cc
+++ b/compiler/optimizing/intrinsics_arm64.cc
@@ -3777,6 +3777,10 @@
   } else {
-    MemOperand field = HeapOperand(WRegisterFrom(obj), referent_offset);
+    UseScratchRegisterScope temps(masm);
+    Register native_obj = temps.AcquireX();
+    __ Cbz(WRegisterFrom(obj), slow_path->GetEntryLabel());
+    ReferenceCodegenARM64<>::DecodeNonNull(masm, native_obj, WRegisterFrom(obj));
+    MemOperand field = HeapOperand(native_obj, referent_offset);
     codegen_->LoadAcquire(
-        invoke, DataType::Type::kReference, WRegisterFrom(out), field, /*needs_null_check=*/ true);
+        invoke, DataType::Type::kReference, WRegisterFrom(out), field, /*needs_null_check=*/ false);
     codegen_->MaybeGenerateReadBarrierSlow(invoke, out, out, obj, referent_offset);
   }
