diff --git a/ojluni/src/main/native/io_util_md.c b/ojluni/src/main/native/io_util_md.c
--- a/ojluni/src/main/native/io_util_md.c
+++ b/ojluni/src/main/native/io_util_md.c
@@ -28,6 +28,13 @@
 #include "io_util.h"
 #include "io_util_md.h"
 #include <string.h>
+#if defined(DARWIN_ART_NATIVEHELPER_FILE_DESCRIPTOR)
+#include <android/file_descriptor_jni.h>
+#endif
 #include <unistd.h>
+#if defined(DARWIN_ART_NATIVEHELPER_FILE_DESCRIPTOR)
+extern int darwin_art_libcore_open(const char*, int, ...)
+    __attribute__((weak_import));
+#endif

 #if defined(__linux__) || defined(_ALLBSD_SOURCE) || defined(_AIX)
@@ -80,7 +87,13 @@
 FD
 handleOpen(const char *path, int oflag, int mode) {
     FD fd;
-    RESTARTABLE(open64(path, oflag, mode), fd);
+#if defined(DARWIN_ART_NATIVEHELPER_FILE_DESCRIPTOR)
+    RESTARTABLE(darwin_art_libcore_open != NULL
+                    ? darwin_art_libcore_open(path, oflag, mode)
+                    : open64(path, oflag, mode), fd);
+#else
+    RESTARTABLE(open64(path, oflag, mode), fd);
+#endif
     if (fd != -1) {
         struct stat64 buf64;
         int result;
@@ -104,7 +117,11 @@ FD getFD(JNIEnv *env, jobject obj, jfieldID fid) {
   if (fdo == NULL) {
     return -1;
   }
+#if defined(DARWIN_ART_NATIVEHELPER_FILE_DESCRIPTOR)
+  return AFileDescriptor_getFd(env, fdo);
+#else
   return (*env)->GetIntField(env, fdo, IO_fd_fdID);
+#endif
 }

 // BEGIN Android-removed: Open files using IoBridge to share BlockGuard & StrictMode logic.
