--- a/ojluni/src/main/native/UNIXProcess_md.c
+++ b/ojluni/src/main/native/UNIXProcess_md.c
@@ -46,7 +46,7 @@
 #include <sys/types.h>
 #include <ctype.h>
 // Android-changed: Fuchsia: Point to correct location of header. http://b/119426171
 // #ifdef _ALLBSD_SOURCE
-#if defined(_ALLBSD_SOURCE) && !defined(__Fuchsia__)
+#if defined(_ALLBSD_SOURCE) && !defined(__Fuchsia__) && !defined(__APPLE__)
 #include <wait.h>
 #else
 #include <sys/wait.h>
