diff --git a/java/android/os/ext/SdkExtensions.java b/java/android/os/ext/SdkExtensions.java
index 2a593c1..0000000 100644
--- a/java/android/os/ext/SdkExtensions.java
+++ b/java/android/os/ext/SdkExtensions.java
@@ -20,0 +21 @@ import android.annotation.NonNull;
+import android.os.Build.VERSION;
@@ -24,2 +24,0 @@ import android.os.SystemProperties;
-import com.android.modules.utils.build.SdkLevel;
-
@@ -66 +64 @@ public class SdkExtensions {
-        if (SdkLevel.isAtLeastS()) {
+        if (VERSION.SDK_INT >= VERSION_CODES.S) {
@@ -69 +67 @@ public class SdkExtensions {
-        if (SdkLevel.isAtLeastT()) {
+        if (VERSION.SDK_INT >= VERSION_CODES.TIRAMISU) {
@@ -73 +71 @@ public class SdkExtensions {
-        if (SdkLevel.isAtLeastU()) {
+        if (VERSION.SDK_INT >= VERSION_CODES.UPSIDE_DOWN_CAKE) {
@@ -76 +74 @@ public class SdkExtensions {
-        if (SdkLevel.isAtLeastV()) {
+        if (VERSION.SDK_INT >= VERSION_CODES.VANILLA_ICE_CREAM) {
@@ -79 +77 @@ public class SdkExtensions {
-        if (SdkLevel.isAtLeastB()) {
+        if (VERSION.SDK_INT >= VERSION_CODES.BAKLAVA) {
