diff --git a/ti_allocator.cc b/ti_allocator.cc
index 5904764..753460e 100644
--- a/ti_allocator.cc
+++ b/ti_allocator.cc
@@ -34,9 +34,9 @@
 #include "ti_allocator.h"
 
 #if defined(__APPLE__)
-// Apple doesn't have malloc.h. Just give this function a non-functional definition.
-#define malloc_usable_size(P) 0
+#include <malloc/malloc.h>
+#define malloc_usable_size(P) malloc_size(P)
 #else
 #include <malloc.h>
 #endif
