diff --git a/pipeline/skia/SkiaOpenGLPipeline.cpp b/pipeline/skia/SkiaOpenGLPipeline.cpp
--- a/pipeline/skia/SkiaOpenGLPipeline.cpp
+++ b/pipeline/skia/SkiaOpenGLPipeline.cpp
@@ -123,14 +123,19 @@ IRenderPipeline::DrawResult SkiaOpenGLPipeline::draw(
         mEglManager.damageFrame(frame, dirty);
     }
 
+#if defined(__APPLE__)
+    SkColorType colorType = kRGBA_8888_SkColorType;
+#else
     SkColorType colorType = getSurfaceColorType();
+#endif
     // setup surface for fbo0
     GrGLFramebufferInfo fboInfo;
     fboInfo.fFBOID = 0;
     if (colorType == kRGBA_F16_SkColorType) {
         fboInfo.fFormat = GL_RGBA16F;
-    } else if (colorType == kN32_SkColorType) {
+    } else if (colorType == kN32_SkColorType ||
+               colorType == kRGBA_8888_SkColorType) {
         // Note: The default preference of pixel format is RGBA_8888, when other
         // pixel format is available, we should branch out and do more check.
